<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<link href="data:text/css;charset=utf-8,%0A%40font%2Dface%20%7B%0Afont%2Dfamily%3A%20octicons%2Dlink%3B%0Asrc%3A%20url%28data%3Afont%2Fwoff%3Bcharset%3Dutf%2D8%3Bbase64%2Cd09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM%2B8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB%2FaFGpk3jaTY6xa8JAGMW%2FO62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v%2Bk%2F0an2i%2BitHDw3v2%2B9%2BDBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3%2FI7AtxEJLtzzuZfI%2BVVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy%2FLt7Kc%2B0vWY%2FgAgIIEqAN9we0pwKXreiMasxvabDQMM4riO%2BqxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw%2Bymhce7vwM9jSqO8JyVd5RH9gyTt2%2BJ%2FyUmYlIR0s04n6%2B7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv%2FocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi%2BW2%2BMjCzMIDApSwvXzC97Z4Ig8N%2FBxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh%2F8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT%2BAEjAwuDFpBmA9KMDEwMCh9i%2Fv8H8sH0%2F4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9%2FlqYwOGZxeUelN2U2R6%2BcArgtCJpauW7UQBqnFkUsjAY%2FkOU1cP%2BDAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl%2BvvmM%2FbyA48e6tWrKArm4ZJlCbdsrxksL1AwWn%2FyBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO%2F%2FsdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd%2F89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF%2B9JOS0nbaaYDCQfwCJ7Au3AHj%2BLO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm%2BEBXuAbHmIMSRMs%2B4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL%2BhD7C1xoaHeLJSEao0FEW14ckxC%2BTU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13%2F%2Blm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl%2B9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O%2FAdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB%2F%2F%2FAA8AAQAAAAAAAAAAAAAAAAABAAAAAA%3D%3D%29%20format%28%27woff%27%29%3B%0A%7D%0Abody%20%7B%0A%2Dwebkit%2Dtext%2Dsize%2Dadjust%3A%20100%25%3B%0Atext%2Dsize%2Dadjust%3A%20100%25%3B%0Acolor%3A%20%23333%3B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20%22Segoe%20UI%22%2C%20Arial%2C%20freesans%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0Afont%2Dsize%3A%2016px%3B%0Aline%2Dheight%3A%201%2E6%3B%0Aword%2Dwrap%3A%20break%2Dword%3B%0A%7D%0Aa%20%7B%0Abackground%2Dcolor%3A%20transparent%3B%0A%7D%0Aa%3Aactive%2C%0Aa%3Ahover%20%7B%0Aoutline%3A%200%3B%0A%7D%0Astrong%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%202em%3B%0Amargin%3A%200%2E67em%200%3B%0A%7D%0Aimg%20%7B%0Aborder%3A%200%3B%0A%7D%0Ahr%20%7B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Aheight%3A%200%3B%0A%7D%0Apre%20%7B%0Aoverflow%3A%20auto%3B%0A%7D%0Acode%2C%0Akbd%2C%0Apre%20%7B%0Afont%2Dfamily%3A%20monospace%2C%20monospace%3B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ainput%20%7B%0Acolor%3A%20inherit%3B%0Afont%3A%20inherit%3B%0Amargin%3A%200%3B%0A%7D%0Ahtml%20input%5Bdisabled%5D%20%7B%0Acursor%3A%20default%3B%0A%7D%0Ainput%20%7B%0Aline%2Dheight%3A%20normal%3B%0A%7D%0Ainput%5Btype%3D%22checkbox%22%5D%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0Apadding%3A%200%3B%0A%7D%0Atable%20%7B%0Aborder%2Dcollapse%3A%20collapse%3B%0Aborder%2Dspacing%3A%200%3B%0A%7D%0Atd%2C%0Ath%20%7B%0Apadding%3A%200%3B%0A%7D%0A%2A%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0A%7D%0Ainput%20%7B%0Afont%3A%2013px%20%2F%201%2E4%20Helvetica%2C%20arial%2C%20nimbussansl%2C%20liberationsans%2C%20freesans%2C%20clean%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%234078c0%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%2C%0Aa%3Aactive%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200%3B%0Amargin%3A%2015px%200%3B%0Aoverflow%3A%20hidden%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23ddd%3B%0A%7D%0Ahr%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ahr%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%2015px%3B%0Amargin%2Dbottom%3A%2015px%3B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%2030px%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%2021px%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%2016px%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%2011px%3B%0A%7D%0Ablockquote%20%7B%0Amargin%3A%200%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Aol%20ol%2C%0Aul%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Droman%3B%0A%7D%0Aul%20ul%20ol%2C%0Aul%20ol%20ol%2C%0Aol%20ul%20ol%2C%0Aol%20ol%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Dalpha%3B%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0Afont%3A%2012px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0A%7D%0A%2Eselect%3A%3A%2Dms%2Dexpand%20%7B%0Aopacity%3A%200%3B%0A%7D%0A%2Eocticon%20%7B%0Afont%3A%20normal%20normal%20normal%2016px%2F1%20octicons%2Dlink%3B%0Adisplay%3A%20inline%2Dblock%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Drendering%3A%20auto%3B%0A%2Dwebkit%2Dfont%2Dsmoothing%3A%20antialiased%3B%0A%2Dmoz%2Dosx%2Dfont%2Dsmoothing%3A%20grayscale%3B%0A%2Dwebkit%2Duser%2Dselect%3A%20none%3B%0A%2Dmoz%2Duser%2Dselect%3A%20none%3B%0A%2Dms%2Duser%2Dselect%3A%20none%3B%0Auser%2Dselect%3A%20none%3B%0A%7D%0A%2Eocticon%2Dlink%3Abefore%20%7B%0Acontent%3A%20%27%5Cf05c%27%3B%0A%7D%0A%2Emarkdown%2Dbody%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%20%21important%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%20%21important%3B%0A%7D%0Aa%3Anot%28%5Bhref%5D%29%20%7B%0Acolor%3A%20inherit%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0A%2Eanchor%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%2Dright%3A%202px%3B%0Amargin%2Dleft%3A%20%2D18px%3B%0A%7D%0A%2Eanchor%3Afocus%20%7B%0Aoutline%3A%20none%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%201em%3B%0Amargin%2Dbottom%3A%2016px%3B%0Afont%2Dweight%3A%20bold%3B%0Aline%2Dheight%3A%201%2E4%3B%0A%7D%0Ah1%20%2Eocticon%2Dlink%2C%0Ah2%20%2Eocticon%2Dlink%2C%0Ah3%20%2Eocticon%2Dlink%2C%0Ah4%20%2Eocticon%2Dlink%2C%0Ah5%20%2Eocticon%2Dlink%2C%0Ah6%20%2Eocticon%2Dlink%20%7B%0Acolor%3A%20%23000%3B%0Avertical%2Dalign%3A%20middle%3B%0Avisibility%3A%20hidden%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%2C%0Ah2%3Ahover%20%2Eanchor%2C%0Ah3%3Ahover%20%2Eanchor%2C%0Ah4%3Ahover%20%2Eanchor%2C%0Ah5%3Ahover%20%2Eanchor%2C%0Ah6%3Ahover%20%2Eanchor%20%7B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah2%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah3%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah4%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah5%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah6%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%20%7B%0Avisibility%3A%20visible%3B%0A%7D%0Ah1%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%202%2E25em%3B%0Aline%2Dheight%3A%201%2E2%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah1%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah2%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%201%2E75em%3B%0Aline%2Dheight%3A%201%2E225%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah2%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%201%2E5em%3B%0Aline%2Dheight%3A%201%2E43%3B%0A%7D%0Ah3%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%201%2E25em%3B%0A%7D%0Ah4%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ah5%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%201em%3B%0Acolor%3A%20%23777%3B%0A%7D%0Ah6%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ap%2C%0Ablockquote%2C%0Aul%2C%0Aol%2C%0Adl%2C%0Atable%2C%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%204px%3B%0Apadding%3A%200%3B%0Amargin%3A%2016px%200%3B%0Abackground%2Dcolor%3A%20%23e7e7e7%3B%0Aborder%3A%200%20none%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%2Dleft%3A%202em%3B%0A%7D%0Aul%20ul%2C%0Aul%20ol%2C%0Aol%20ol%2C%0Aol%20ul%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Ali%3Ep%20%7B%0Amargin%2Dtop%3A%2016px%3B%0A%7D%0Adl%20%7B%0Apadding%3A%200%3B%0A%7D%0Adl%20dt%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%2016px%3B%0Afont%2Dsize%3A%201em%3B%0Afont%2Dstyle%3A%20italic%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Adl%20dd%20%7B%0Apadding%3A%200%2016px%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ablockquote%20%7B%0Apadding%3A%200%2015px%3B%0Acolor%3A%20%23777%3B%0Aborder%2Dleft%3A%204px%20solid%20%23ddd%3B%0A%7D%0Ablockquote%3E%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Ablockquote%3E%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Atable%20%7B%0Adisplay%3A%20block%3B%0Awidth%3A%20100%25%3B%0Aoverflow%3A%20auto%3B%0Aword%2Dbreak%3A%20normal%3B%0Aword%2Dbreak%3A%20keep%2Dall%3B%0A%7D%0Atable%20th%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Atable%20th%2C%0Atable%20td%20%7B%0Apadding%3A%206px%2013px%3B%0Aborder%3A%201px%20solid%20%23ddd%3B%0A%7D%0Atable%20tr%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Aborder%2Dtop%3A%201px%20solid%20%23ccc%3B%0A%7D%0Atable%20tr%3Anth%2Dchild%282n%29%20%7B%0Abackground%2Dcolor%3A%20%23f8f8f8%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Abackground%2Dcolor%3A%20%23fff%3B%0A%7D%0Acode%20%7B%0Apadding%3A%200%3B%0Apadding%2Dtop%3A%200%2E2em%3B%0Apadding%2Dbottom%3A%200%2E2em%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%2085%25%3B%0Abackground%2Dcolor%3A%20rgba%280%2C0%2C0%2C0%2E04%29%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0Acode%3Abefore%2C%0Acode%3Aafter%20%7B%0Aletter%2Dspacing%3A%20%2D0%2E2em%3B%0Acontent%3A%20%22%5C00a0%22%3B%0A%7D%0Apre%3Ecode%20%7B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%20100%25%3B%0Aword%2Dbreak%3A%20normal%3B%0Awhite%2Dspace%3A%20pre%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0A%2Ehighlight%20%7B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0A%2Ehighlight%20pre%2C%0Apre%20%7B%0Apadding%3A%2016px%3B%0Aoverflow%3A%20auto%3B%0Afont%2Dsize%3A%2085%25%3B%0Aline%2Dheight%3A%201%2E45%3B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0A%2Ehighlight%20pre%20%7B%0Amargin%2Dbottom%3A%200%3B%0Aword%2Dbreak%3A%20normal%3B%0A%7D%0Apre%20%7B%0Aword%2Dwrap%3A%20normal%3B%0A%7D%0Apre%20code%20%7B%0Adisplay%3A%20inline%3B%0Amax%2Dwidth%3A%20initial%3B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Aoverflow%3A%20initial%3B%0Aline%2Dheight%3A%20inherit%3B%0Aword%2Dwrap%3A%20normal%3B%0Abackground%2Dcolor%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0Apre%20code%3Abefore%2C%0Apre%20code%3Aafter%20%7B%0Acontent%3A%20normal%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%2Dsize%3A%2011px%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Epl%2Dc%20%7B%0Acolor%3A%20%23969896%3B%0A%7D%0A%2Epl%2Dc1%2C%0A%2Epl%2Ds%20%2Epl%2Dv%20%7B%0Acolor%3A%20%230086b3%3B%0A%7D%0A%2Epl%2De%2C%0A%2Epl%2Den%20%7B%0Acolor%3A%20%23795da3%3B%0A%7D%0A%2Epl%2Ds%20%2Epl%2Ds1%2C%0A%2Epl%2Dsmi%20%7B%0Acolor%3A%20%23333%3B%0A%7D%0A%2Epl%2Dent%20%7B%0Acolor%3A%20%2363a35c%3B%0A%7D%0A%2Epl%2Dk%20%7B%0Acolor%3A%20%23a71d5d%3B%0A%7D%0A%2Epl%2Dpds%2C%0A%2Epl%2Ds%2C%0A%2Epl%2Ds%20%2Epl%2Dpse%20%2Epl%2Ds1%2C%0A%2Epl%2Dsr%2C%0A%2Epl%2Dsr%20%2Epl%2Dcce%2C%0A%2Epl%2Dsr%20%2Epl%2Dsra%2C%0A%2Epl%2Dsr%20%2Epl%2Dsre%20%7B%0Acolor%3A%20%23183691%3B%0A%7D%0A%2Epl%2Dv%20%7B%0Acolor%3A%20%23ed6a43%3B%0A%7D%0A%2Epl%2Did%20%7B%0Acolor%3A%20%23b52a1d%3B%0A%7D%0A%2Epl%2Dii%20%7B%0Abackground%2Dcolor%3A%20%23b52a1d%3B%0Acolor%3A%20%23f8f8f8%3B%0A%7D%0A%2Epl%2Dsr%20%2Epl%2Dcce%20%7B%0Acolor%3A%20%2363a35c%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dml%20%7B%0Acolor%3A%20%23693a17%3B%0A%7D%0A%2Epl%2Dmh%2C%0A%2Epl%2Dmh%20%2Epl%2Den%2C%0A%2Epl%2Dms%20%7B%0Acolor%3A%20%231d3e81%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmq%20%7B%0Acolor%3A%20%23008080%3B%0A%7D%0A%2Epl%2Dmi%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dstyle%3A%20italic%3B%0A%7D%0A%2Epl%2Dmb%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmd%20%7B%0Abackground%2Dcolor%3A%20%23ffecec%3B%0Acolor%3A%20%23bd2c00%3B%0A%7D%0A%2Epl%2Dmi1%20%7B%0Abackground%2Dcolor%3A%20%23eaffea%3B%0Acolor%3A%20%2355a532%3B%0A%7D%0A%2Epl%2Dmdr%20%7B%0Acolor%3A%20%23795da3%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmo%20%7B%0Acolor%3A%20%231d3e81%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%3A%2011px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20%7B%0Alist%2Dstyle%2Dtype%3A%20none%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%2B%2Etask%2Dlist%2Ditem%20%7B%0Amargin%2Dtop%3A%203px%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20input%20%7B%0Amargin%3A%200%200%2E35em%200%2E25em%20%2D1%2E6em%3B%0Avertical%2Dalign%3A%20middle%3B%0A%7D%0A%3Achecked%2B%2Eradio%2Dlabel%20%7B%0Az%2Dindex%3A%201%3B%0Aposition%3A%20relative%3B%0Aborder%2Dcolor%3A%20%234078c0%3B%0A%7D%0Acode%20%3E%20%2Ekw%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Edt%20%7B%20color%3A%20%23ed6a43%3B%20%7D%0Acode%20%3E%20%2Edv%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ebn%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Efl%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ech%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Est%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eco%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Eot%20%7B%20color%3A%20%230086b3%3B%20%7D%0Acode%20%3E%20%2Eal%20%7B%20color%3A%20%23a61717%3B%20%7D%0Acode%20%3E%20%2Efu%20%7B%20color%3A%20%2363a35c%3B%20%7D%0Acode%20%3E%20%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%0Acode%20%3E%20%2Ewa%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ecn%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Esc%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Evs%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Ess%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eim%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eva%20%7Bcolor%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecf%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eop%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ebu%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eex%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Epp%20%7B%20color%3A%20%23999999%3B%20%7D%0Acode%20%3E%20%2Eat%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Edo%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Ean%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecv%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ein%20%7B%20color%3A%20%23008080%3B%20%7D%0A" rel="stylesheet">
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="taxonomic-backbone-for-splot-2.1-and-try-3.0">Taxonomic backbone for sPlot 2.1 and TRY 3.0</h1>
<p>Oliver Purschke<br />
12 August, 2017</p>
<h1 id="combine-taxon-names-from-splot-2.0-and-try-3.0">Combine taxon names from sPlot 2.0 and TRY 3.0</h1>
<h2 id="load-required-packages">Load required packages</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(stringr)
<span class="kw">library</span>(knitr)
<span class="kw">library</span>(vegdata)
<span class="kw">library</span>(doParallel)
<span class="kw">library</span>(dplyr)
<span class="kw">library</span>(plyr)
<span class="kw">library</span>(Taxonstand)
<span class="kw">library</span>(rgbif)</code></pre></div>
<h2 id="read-in-taxon-names-from-splot-and-try">Read in taxon names from <a href="https://www.idiv.de/sdiv/working_groups/wg_pool/splot/splot_database.html">sPlot</a> and <a href="https://www.try-db.org/TryWeb/Home.php">TRY</a></h2>
<h3 id="reading-splot-species">Reading sPlot Species</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path.sPlot &lt;-<span class="st"> &quot;/))</span>
<span class="st">home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/Analyses/Data/Species/</span>
<span class="st">sPlot/sPlot_14_04_2015/&quot;</span>
sPlot.version &lt;-<span class="st"> &quot;sPlot_14_4_2015_species&quot;</span>

splot.species &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">paste0</span>(path.sPlot, sPlot.version, <span class="st">&quot;.csv&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">gc</span>()</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(splot.species)</code></pre></div>
<pre><code>##   PlotObservationID Taxonomy Taxon.group Taxon.group.ID
## 1                 1 Pasl2012     Unknown              0
## 2                 1 Pasl2012     Unknown              0
## 3                 1 Pasl2012     Unknown              0
## 4                 1 Pasl2012     Unknown              0
## 5                 1 Pasl2012     Unknown              0
## 6                 1 Pasl2012     Unknown              0
##                Turboveg2.concept                Matched.concept Match
## 1     Allantoparmelia almquistii     Allantoparmelia almquistii     0
## 2             Andreaea rupestris             Andreaea rupestris     0
## 3       Arctoparmelia centrifuga       Arctoparmelia centrifuga     0
## 4             Cetraria nigricans             Cetraria nigricans     0
## 5           Cladonia amaurocraea           Cladonia amaurocraea     0
## 6 Cladonia arbuscula subsp. lat. Cladonia arbuscula subsp. lat.     0
##   Original.taxon.concept Layer Cover.. Cover.code x_
## 1                            0       1          1 NA
## 2                            0       2          2 NA
## 3                            0       5          5 NA
## 4                            0       2          2 NA
## 5                            0       1          1 NA
## 6                            0       2          2 NA</code></pre>
<p>Use the <code>Matched.concept</code> column (Federhen 2010), as it already contains some standardization by <a href="https://www.wur.nl/de/Personen/drs.-SM-Stephan-Hennekens.htm">Stephan Hennekkens</a> according to <a href="https://www.synbiosys.alterra.nl/turboveg/contact.htm">synbiosys</a>[1].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splot.species$Matched.concept&lt;-<span class="st"> </span><span class="kw">as.character</span>(splot.species$Matched.concept)</code></pre></div>
<h3 id="read-in-try-3.0-species-list">Read in TRY 3.0 species list</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">try3.species &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/Analyses/Data/</span>
<span class="st">Species/TRY/Species/TRY30_Gapfilling2015_Species.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(try3.species)</code></pre></div>
<pre><code>##   AccSpeciesID               Species
## 1            1                 Aa sp
## 2            2    Abarema adenophora
## 3            3   Abarema barbouriana
## 4            4 Abarema brachystachya
## 5            6       Abarema jupunba
## 6            7         Abarema laeta</code></pre>
<h2 id="combine-species-lists">Combine species lists</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(<span class="kw">c</span>(<span class="kw">as.character</span>(splot.species$Matched.concept),
                                     <span class="kw">as.character</span>(try3.species$Species))))
<span class="kw">length</span>(spec.list.TRY.sPlot)</code></pre></div>
<p>Create dataframe that will later hold the original (uncleaned) names in sPlot and TRY:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">cbind</span>(spec.list.TRY.sPlot, spec.list.TRY.sPlot,
                               spec.list.TRY.sPlot)
<span class="kw">str</span>(spec.list.TRY.sPlot<span class="fl">.2</span>)
<span class="kw">dim</span>(spec.list.TRY.sPlot<span class="fl">.2</span>)
spec.list.TRY.sPlot<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(spec.list.TRY.sPlot<span class="fl">.2</span>)</code></pre></div>
<p>Convert factors into characters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">as.character</span>(spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">1</span>])
spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">as.character</span>(spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">2</span>])
spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">as.character</span>(spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="dv">3</span>])</code></pre></div>
<p>Give column names[2]:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(spec.list.TRY.sPlot<span class="fl">.2</span>) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;names.sPlot.TRY&quot;</span>, <span class="st">&quot;names.corr.string&quot;</span>, <span class="st">&quot;sPlot.TRY&quot;</span>)
<span class="kw">write.csv</span>(spec.list.TRY.sPlot<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;spec.list.TRY3.sPlot2.csv&quot;</span>)</code></pre></div>
<h2 id="add-species-database-affiliation">Add species' database affiliation</h2>
<p>Species in sPlot 2.0:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot<span class="fl">.2</span>$sPlot.TRY[<span class="kw">which</span>(spec.list.TRY.sPlot<span class="fl">.2</span>$names.sPlot.TRY %in%
<span class="st">                                      </span><span class="kw">unique</span>(splot.species$Matched.concept))] &lt;-<span class="st"> &quot;S&quot;</span></code></pre></div>
<p>Species in TRY 3.0:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot<span class="fl">.2</span>$sPlot.TRY[<span class="kw">which</span>(spec.list.TRY.sPlot<span class="fl">.2</span>$names.sPlot.TRY %in%
<span class="st">                                      </span><span class="kw">unique</span>(try3.species$Species))] &lt;-<span class="st"> &quot;T&quot;</span></code></pre></div>
<p>Species that are both in sPlot 2.0 and TRY 3.0</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spec.list.TRY.sPlot<span class="fl">.2</span>$sPlot.TRY[<span class="kw">which</span>(spec.list.TRY.sPlot<span class="fl">.2</span>$names.sPlot.TRY %in%
<span class="st">                                      </span><span class="kw">unique</span>(try3.species$Species) &amp;
<span class="st">                                      </span>spec.list.TRY.sPlot<span class="fl">.2</span>$names.sPlot.TRY %in%
<span class="st">                                      </span><span class="kw">unique</span>(splot.species$Matched.concept))] &lt;-<span class="st"> &quot;ST&quot;</span></code></pre></div>
<h1 id="a-priori-cleaning-of-names">A-priori cleaning of names</h1>
<h2 id="manual-cleaning">Manual cleaning</h2>
<p>Based one the column <code>Matched.concept</code> in <code>sPlot</code>, a list of 4,093 &quot;weird&quot; species names (consisting mainly of trivial names) was generated, and corrected manually by <a href="https://www.uni-bayreuth.de/de/forschung/profilfelder/advanced-fields/oekologie-und-umweltwissenschaften/mitwirkende/Dengler/index.php">Jürgen Dengler</a> (thereafter JD). Some examples:</p>
<ul>
<li>Gras silbrig Bl haarig 133106 → Poaceae sp. [silbrig Bl haarig 133106]</li>
<li>Ãƒâ€”Achnella species → Achnella sp.</li>
<li>[KHH Composite (breite Bl.)] → Asteraceae sp. [breite Bl.]</li>
<li>LICH Xanthomaculina hottentotta → Xanthomaculina hottentotta</li>
<li>cf. Silberknospe 134249 → Spermatophyta sp. [Silberknospe 134249]</li>
</ul>
<h2 id="string-manipulation-routines">String manipulation routines</h2>
<p>Stripping unwanted characters as well as abbreviation (such as hybrid markers) which would prevent name matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'*'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'cf. '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'Cf. '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'['</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">']'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' x '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'Ã—'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'aff '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'('</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">')'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' cf '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' aff. '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'c‚e'</span>, <span class="st">'ceae'</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'    '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'   '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'  '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'x-'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'X-'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'like '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)  
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">','</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)  </code></pre></div>
<p>For all names, that have a number in their first word, and consist of &gt; 1 words, remove that word:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(stringr)
firstWordWithNumbers &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">'[0-9]'</span>, <span class="kw">word</span>(OriginalNames, <span class="dv">1</span>))
numberOfWords &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">gregexpr</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">W+&quot;</span>, OriginalNames), length) +<span class="st"> </span><span class="dv">1</span>
OriginalNames[firstWordWithNumbers &amp;<span class="st"> </span>numberOfWords &gt;<span class="st"> </span><span class="dv">1</span>] &lt;-
<span class="st">    </span><span class="kw">sapply</span>(OriginalNames[firstWordWithNumbers &amp;<span class="st"> </span>numberOfWords &gt;<span class="st"> </span><span class="dv">1</span>],
           function(x) <span class="kw">substr</span>(x, <span class="dt">start=</span><span class="kw">regexpr</span>(<span class="dt">pattern =</span><span class="st">' '</span>, <span class="dt">text=</span>x)+<span class="dv">1</span>, <span class="dt">stop=</span><span class="kw">nchar</span>(x)))</code></pre></div>
<p>Correct some name abbreviations using <code>taxname.abbr</code> in <code>vegdata</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">CleanedNames &lt;-<span class="st"> </span><span class="kw">taxname.abbr</span>(OriginalNames)
<span class="kw">save</span>(CleanedNames, <span class="dt">file =</span> <span class="st">&quot;CleanedNames.Rdata&quot;</span>)
<span class="kw">write.csv</span>(CleanedNames, <span class="dt">file =</span> <span class="st">&quot;CleanedNames.csv&quot;</span>)
<span class="kw">length</span>(CleanedNames)</code></pre></div>
<p>Create a data frame from the string cleaned names:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">CleanedNames.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(CleanedNames, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
spec.list.TRY.sPlot<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">cbind</span>(spec.list.TRY.sPlot<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>)], CleanedNames.df,
                               CleanedNames.df)</code></pre></div>
<p>Skip column <code>names.corr.string</code>, as the <code>cleanednames.df</code> column as well as another column that will include the names corrected by JD.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(spec.list.TRY.sPlot<span class="fl">.3</span>) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;index&quot;</span>,<span class="st">&quot;original.names.sPlot.TRY&quot;</span>,<span class="st">&quot;sPlot.TRY&quot;</span>,
                                     <span class="st">&quot;CleanedNames&quot;</span>,<span class="st">&quot;CleanedNames.Juergen&quot;</span>)
Tax_Back_sPlot2_TRY3 &lt;-<span class="st"> </span>spec.list.TRY.sPlot<span class="fl">.3</span>
<span class="kw">save</span>(Tax_Back_sPlot2_TRY3, <span class="dt">file =</span> <span class="st">&quot;Tax_Back_sPlot2_TRY3.Rdata&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;Tax_Back_sPlot2_TRY3.Rdata&quot;</span>)
<span class="kw">write.csv</span>(Tax_Back_sPlot2_TRY3, <span class="dt">file =</span> <span class="st">&quot;Tax_Back_sPlot2_TRY3.csv&quot;</span>)
<span class="kw">head</span>(Tax_Back_sPlot2_TRY3)</code></pre></div>
<h3 id="substitute-a-priorily-cleaned-names-with-corrected-names">Substitute a-priorily cleaned names with corrected names</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">names.correct.JD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;weird.names_JD_string_correct.csv&quot;</span>)
<span class="kw">str</span>(names.correct.JD)
<span class="kw">head</span>(names.correct.JD)
index &lt;-<span class="st"> </span><span class="kw">match</span>(names.correct.JD$weird.names, Tax_Back_sPlot2_TRY3$original.names.sPlot.TRY)
Tax_Back_sPlot2_TRY3$CleanedNames.JD[index] &lt;-<span class="st"> </span>names.correct.JD$Name.corrected</code></pre></div>
<h3 id="do-some-more-cleaning-on-cleanednames.jd">Do some more cleaning on <code>CleanedNames.JD</code></h3>
<p>This was done manually in a csv-file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">CleanedNames.JD &lt;-<span class="st"> </span>Tax_Back_sPlot2_TRY3$CleanedNames.JD
<span class="kw">write.csv</span>(CleanedNames.JD, <span class="dt">file =</span> <span class="st">&quot;CleanedNames.JD.csv&quot;</span>)</code></pre></div>
<h3 id="substitute-old-cleanednames.jd-with-new-string-manipulated-one">Substitute old <code>CleanedNames.JD</code> with new string-manipulated one:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Tax_Back_sPlot2_TRY3$CleanedNames.JD &lt;-<span class="st"> </span>CleanedNames.JD
<span class="kw">save</span>(Tax_Back_sPlot2_TRY3, <span class="dt">file =</span> <span class="st">&quot;Tax_Back_sPlot2_TRY3.Rdata&quot;</span>)</code></pre></div>
<h1 id="match-names-against-taxonomic-name-resolution-service-tnrs">Match names against Taxonomic Name Resolution Service (<a href="http://tnrs.iplantcollaborative.org">TNRS</a>)</h1>
<h2 id="slice-cleanednames-into-chunks">Slice <code>CleanedNames</code> into chunks</h2>
<p>... of 5000 species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seq1 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from =</span><span class="dv">1</span>, <span class="dt">to =</span> <span class="dv">120001</span>, <span class="dv">5000</span>)
seq2 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from =</span><span class="dv">5000</span>, <span class="dt">to =</span> <span class="dv">125000</span>, <span class="dv">5000</span>)

for(i in <span class="dv">1</span>:<span class="kw">length</span>(seq1)) {
    <span class="kw">write.csv</span>(Tax_Back_sPlot2_TRY3$CleanedNames.JD[seq1[i]:seq2[i]],
              <span class="dt">file =</span> <span class="kw">paste</span>(<span class="kw">paste</span>(<span class="st">&quot;tnrs_submit&quot;</span>, seq1[i], <span class="dt">sep =</span> <span class="st">&quot;_&quot;</span>), <span class="st">&quot;csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>))
}</code></pre></div>
<p>The single csv-files, containing 5,000 names each, were submitted to <a href="http://tnrs.iplantcollaborative.org/sources.html">Taxonomic Name Resolution Service web application</a> (Boyle <em>et al.</em> 2013, iPlant Collaborative (2015)). TNRS version 4.0 was used, which became available in August 2015 (this version also included The Plant List version 1.1).</p>
<h2 id="tnrs-settings">TNRS settings</h2>
<p>The following settings were used for resolving names on TNRS.</p>
<h3 id="sources-for-name-resolution">Sources for name resolution</h3>
<p>The initial TNRS name resolution run was based on the <strong>five standard sources</strong> that were <strong>ranked according to preference</strong> in the following order (default of TNRS):</p>
<ol>
<li>The Plant List (<a href="http://www.theplantlist.org/">TPL</a>)(The Plant List 2013)</li>
<li>The Global Compositae Checklist (<a href="http://compositae.landcareresearch.co.nz/">GCC</a>)(Flann 2009)</li>
<li>The International Legume Database and Information Service (<a href="http://www.ildis.org/">ILDIS</a>)(International legume database and information service 2006)</li>
<li><a href="http://www.tropicos.org/">Tropicos</a> (Missouri Botanical Garden 2013)</li>
<li>PLANTS Database (<a href="https://plants.usda.gov/java/">USDA</a>)(USDA, NRCS 2012)</li>
</ol>
<p>Because it is possible that the best match is found in lower ranked sources, see section <a href="#ID">TNRS settings</a>, two additional name resolution runs were realized in which the highest ranking was given to <strong>(1) Tropicos</strong>, or <strong>(2)</strong> the sixth source available in TNRS, <strong>NCBI</strong> (The National Center for Biotechnology Information's Taxonomy database; (Federhen 2010)), respectively, see section <a href="#ID">TNRS settings</a>.</p>
<h3 id="family-classification">Family Classification</h3>
<p>Resolved names were assigned to families based on the <a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1095-8339.2009.00996.x/abstract">APGIII classification</a> (Chase &amp; Reveal 2009), the same classification system used by Tropicos.</p>
<h2 id="retrieve-results">Retrieve results</h2>
<p>Once the matching process was finished, results were retrieved from TNRS using the <code>Detailed Download</code> option that included the full name information (parsed components, warnings, links to sources, etc.). We retrived the single best match for each species, constrained by source (TNRS default), where the name in the first source was selected as best match, unless there was <code>no suitable match found</code> in that source, the match from the next lower-ranked source was selected, until all resources where exhausted.</p>
<h1 id="manually-inspecting-name-matching-results">Manually inspecting name matching results</h1>
<p>Manually inspect the TNRS-results table in a spreadsheat application (i.e. LibreOffice or Excel). Starting with the highest taxonomic rank considered (i.e. Family). For instance, if manual checking of the TRNS output reveals that all accepted names or synonyms that have accuracy scores &gt;0.9 are correct taxon names, use the following selection procedure:</p>
<ul>
<li>Name_matched_rank (==Family)</li>
<li>Taxonomic_status (==Accepted, Synomyn)</li>
<li>Family_score (&gt;0.9)</li>
</ul>
<p>Continue this selection procedure for entries that were matched at lower taxonomic ranks, i.e. genus, species, etc..</p>
<h2 id="read-and-combine-tnrs-result-files">Read and combine TNRS result files</h2>
<h3 id="set-up-doparallel-cluster-to-speed-up-the-reading-process">Set up <code>doParallel</code> cluster to speed up the reading process</h3>
<p>Find out how many cores are available:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">detectCores</span>()</code></pre></div>
<pre><code>## [1] 4</code></pre>
<p>Create cluster with desired number of cores:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cl &lt;-<span class="st"> </span><span class="kw">makeCluster</span>(<span class="dv">3</span>)</code></pre></div>
<p>Register cluster:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">registerDoParallel</span>(cl)</code></pre></div>
<p>Find out how many cores are being used:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getDoParWorkers</span>()</code></pre></div>
<h3 id="read-and-combine-the-single-files">Read and combine the single files</h3>
<p>Read the downloaded TNRS files, including 5000 names each[3] into <code>R</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setwd</span>(<span class="st">&quot;/home/oliver/Downloads/&quot;</span>)
seq1 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dt">from =</span><span class="dv">1</span>, <span class="dt">to =</span> <span class="dv">120001</span>, <span class="dv">5000</span>)
<span class="kw">system.time</span>(
    x &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span>:<span class="kw">length</span>(seq1), <span class="dt">.combine =</span> rbind) %dopar%<span class="st"> </span>{
        <span class="kw">read.csv</span>(<span class="kw">paste</span>(<span class="kw">paste</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.tpl&quot;</span>, seq1[i], <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>),
                       <span class="st">&quot;csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>, <span class="dt">skip =</span> <span class="dv">0</span>,
                 <span class="dt">head =</span> T)[-<span class="dv">1</span>, ]
    }
)
tnrs.tpl &lt;-<span class="st"> </span>x</code></pre></div>
<p>Assign index to first column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl$Name_number &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="kw">length</span>(tnrs.tpl$Name_number)</code></pre></div>
<h2 id="select-correctly-resolved-names">Select correctly resolved names</h2>
<h3 id="general-procedure">General procedure</h3>
<ol>
<li>Open <code>tnrs.tpl</code> in a spread sheat program and sort according to <code>Name_matched_rank</code>, <code>Taxonomic_status</code> and <code>Family_score</code>[4].</li>
<li><p>Repeat selection for entries matched at lower taxonomic ranks, such <code>Name_matched_rank</code> ==:</p>
<ul>
<li>forma</li>
<li>genus</li>
<li>infraspecies</li>
<li>...</li>
</ul></li>
<li><p>Adjust accuracy score threshold values, e.g. use higher or lower values for infraspec., variety, ...</p></li>
</ol>
<h3 id="family-level">Family level</h3>
<p>Manually inspect sorted table and select all entries at the highest hierarchical level (family). Manually identify the family accuracy score threshold value above which a name can be considered a correct name. In the following case, this corresponds to a score $&gt;$0.88.</p>
<h4 id="create-index-for-correct-names">Create index for correct names</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.family &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;family&quot;</span> &amp;
<span class="st">                               </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                </span>tnrs.tpl$Family_score &gt;<span class="st"> </span><span class="fl">0.88</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.family)</code></pre></div>
<h3 id="forma-level">Forma level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.forma &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;forma&quot;</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.forma)</code></pre></div>
<h3 id="genus-level">Genus level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.genus &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                              </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> &amp;
<span class="st">                              </span>tnrs.tpl$Genus_score &gt;<span class="st"> </span><span class="fl">0.83</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus)</code></pre></div>
<h3 id="infraspecies-level">Infraspecies level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.infraspec &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;infraspecies&quot;</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.infraspec)</code></pre></div>
<h3 id="species-level">Species level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.species &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                 </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                  </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                  </span>tnrs.tpl$Genus_score &gt;<span class="st"> </span><span class="fl">0.78</span> &amp;
<span class="st">                                  </span>tnrs.tpl$Name_score &gt;<span class="st"> </span><span class="fl">0.93</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species)</code></pre></div>
<h3 id="subspecies-level">Subspecies level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.subspec &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.subspec)</code></pre></div>
<h3 id="variety-level">Variety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.variety &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.variety)</code></pre></div>
<h3 id="identifying-non-matched-species-that-are-spermatophyta">Identifying &quot;non-matched&quot; species that are spermatophyta</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.spermatophyt &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span>
                                     &amp;<span class="st"> </span><span class="kw">word</span>(tnrs.tpl$Name_submitted, <span class="dv">1</span>) ==<span class="st"> &quot;Spermatophyta&quot;</span>)
                             , <span class="dv">1</span>]
<span class="kw">length</span>(index.spermatophyt)</code></pre></div>
<h2 id="select-certain-or-uncertain-names">Select <code>certain</code> or <code>uncertain</code> names</h2>
<p>Select names that do not fulfill the search criteria, i.e. that were not selected as certain species, for further name matching.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.tpl &lt;-<span class="st"> </span><span class="kw">c</span>(index.family, index.forma, index.genus, index.species, index.subspec,
               index.variety, index.spermatophyt)
<span class="kw">length</span>((index.tpl))

tnrs.tpl.certain &lt;-<span class="st"> </span>tnrs.tpl[index.tpl,]
<span class="kw">dim</span>(tnrs.tpl.certain)
<span class="kw">save</span>(tnrs.tpl.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.certain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.tpl.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.certain.csv&quot;</span>)

tnrs.tpl.uncertain &lt;-<span class="st"> </span>tnrs.tpl[tnrs.tpl$Name_number %in%<span class="st"> </span>index.tpl ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(tnrs.tpl.uncertain)
<span class="kw">save</span>(tnrs.tpl.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.tpl.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.csv&quot;</span>)</code></pre></div>
<p><strong>Generate list of <code>uncertain</code> species that are still to be resolved on TNRS:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(tnrs.tpl.uncertain[,<span class="dv">2</span>], <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.upload.csv&quot;</span>)</code></pre></div>
<h2 id="resolve-uncertain-names-using-tropicos">Resolve 'uncertain' names using <code>Tropicos</code></h2>
<p>Because the matching procedure <a href="Sources%20for%20name%20resolution">above</a>, giving the highest ranking to the sources TPL, GCC and ILDIS (which seem to be the most reliable and up-to-date), will not always result in the match, in cases where better matching scores are achieved in lower ranked sources. Therefore the TNRS matching procedure was continued for the <code>uncertain</code> species from the <a href="Select%20correctly%20resolved%20names">previous step</a>, but assigning the highest rank to the source <code>Tropicos</code>, while repeating the steps that were used to generate <code>tnrs.tpl</code>.[5]</p>
<h3 id="read-in-tnrs-tropicos-results">Read in 'TNRS-Tropicos' results</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setwd</span>(<span class="st">&quot;/home/oliver/Downloads/&quot;</span>)

<span class="kw">system.time</span>(
    x &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span>:<span class="kw">length</span>(seq1), <span class="dt">.combine =</span> rbind) %dopar%<span class="st"> </span>{
        <span class="kw">read.csv</span>(<span class="kw">paste</span>(<span class="kw">paste</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.trop&quot;</span>, seq1[i], <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>),
                       <span class="st">&quot;csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>, <span class="dt">skip =</span> <span class="dv">0</span>,
                 <span class="dt">head =</span> T)[-<span class="dv">1</span>, ]
    }
)

tnrs.trop &lt;-<span class="st"> </span>x

<span class="kw">str</span>(tnrs.trop)
tnrs.trop$Name_number &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="kw">length</span>(tnrs.trop$Name_number)</code></pre></div>
<h3 id="reduce-to-set-of-uncertain-species-tnrs.tpl">Reduce to set of uncertain species <code>tnrs.tpl</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop.small &lt;-<span class="st"> </span>tnrs.trop[tnrs.all.small.uncertain$Name_number, ]
<span class="kw">str</span>(tnrs.trop.small)

<span class="kw">save</span>(tnrs.trop.small, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.csv&quot;</span>)</code></pre></div>
<h2 id="selecting-correctly-resolved-names">Selecting correctly resolved names</h2>
<p>Using the same procedure as <a href="Family%20level">above</a>.</p>
<h3 id="family-level-1">Family level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.family &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;family&quot;</span> &amp;
<span class="st">                                      </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span>|
<span class="st">                                       </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.family)</code></pre></div>
<h3 id="forma-level-1">Forma level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.forma &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;forma&quot;</span> &amp;
<span class="st">                                     </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                      </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.forma)</code></pre></div>
<h3 id="genus-level-1">Genus level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.genus &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                                     </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                      </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                      </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.83</span> &amp;
<span class="st">                                      </span>tnrs.trop.small$Name_score &gt;<span class="st"> </span><span class="fl">0.5</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus)</code></pre></div>
<h3 id="species-level-1">Species level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.species1 &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                        </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                         </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                         </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.88</span> &amp;
<span class="st">                                         </span>tnrs.trop.small$Name_score &gt;<span class="st"> </span><span class="fl">0.9</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species1)

index.species2 &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                        </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                         </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                         </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.78</span> &amp;
<span class="st">                                         </span>tnrs.trop.small$Name_score &gt;<span class="st"> </span><span class="fl">0.94</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species1)

index.species3 &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                        </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                         </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                        </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.88</span> &amp;
<span class="st">                                        </span>tnrs.trop.small$Name_score &gt;<span class="st"> </span><span class="fl">0.49</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species3)

index.species &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.species1, index.species2, index.species3))
<span class="kw">length</span>(index.species)</code></pre></div>
<h3 id="subspecies-level-1">Subspecies level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.subspec &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span> &amp;
<span class="st">                                       </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                        </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.subspec)</code></pre></div>
<h3 id="variety-level-1">Variety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.variety &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span> &amp;
<span class="st">                                       </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                        </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.variety)</code></pre></div>
<h3 id="select-certain-and-uncertain-names">Select <code>certain</code> and <code>uncertain</code> names</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.trop &lt;-<span class="st"> </span><span class="kw">c</span>(index.family, index.forma, index.genus, index.species, index.subspec,
                index.variety)
<span class="kw">length</span>((index.trop))
tnrs.trop.small.certain &lt;-<span class="st"> </span>tnrs.trop.small[tnrs.trop.small$Name_number %in%<span class="st"> </span>index.trop
                                           ==<span class="st"> </span>T,]
<span class="kw">dim</span>(tnrs.trop.small.certain)
<span class="kw">save</span>(tnrs.trop.small.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.certain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.certain.csv&quot;</span>)

tnrs.trop.small.uncertain &lt;-<span class="st"> </span>tnrs.trop.small[tnrs.trop.small$Name_number %in%<span class="st"> </span>index.trop
                                             ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(tnrs.trop.small.uncertain)
<span class="kw">save</span>(tnrs.trop.small.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.uncertain.csv&quot;</span>)</code></pre></div>
<h2 id="resolve-uncertain-names-using-ncbi">Resolve ‘uncertain’ names using <code>NCBI</code></h2>
<p>Cut the list of the remaining 9,641 unresolved species into chunks of ∼ 5,000 species.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(tnrs.trop.small.uncertain$Name_submitted[<span class="dv">1</span>:<span class="dv">5000</span>], <span class="dt">file =</span> <span class="st">&quot;trop.uncert.1.csv&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small.uncertain$Name_submitted[<span class="dv">5001</span>:<span class="dv">9641</span>], <span class="dt">file =</span> <span class="st">&quot;trop.uncert.5001.csv&quot;</span>)</code></pre></div>
<p>Match these lists against TNRS using the settings decribed <a href="TNRS%20settings">above</a>, but rank the additional (sixth) sources in TNRS, The National Center for Biotechnology Information's Taxonomy database <a href>NCBI</a>, first.</p>
<h3 id="read-in-tnrs-ncbi-results">Read in ‘TNRS-NCBI’ results</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.trop.uncert.1.csv&quot;</span>,
                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)[-<span class="dv">1</span>,]
tnrs.ncbi<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.trop.uncert.5001.csv&quot;</span>,
                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)[-<span class="dv">1</span>,]

tnrs.ncbi &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.ncbi<span class="fl">.1</span>, tnrs.ncbi<span class="fl">.2</span>)
<span class="kw">str</span>(tnrs.ncbi)
tnrs.ncbi$Name_number &lt;-<span class="st"> </span>tnrs.trop.small.uncertain$Name_number
<span class="kw">range</span>(tnrs.ncbi$Name_number)

<span class="kw">save</span>(tnrs.ncbi, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.ncbi, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.csv&quot;</span>)</code></pre></div>
<h2 id="selecting-correctly-resolved-names-in-tnrs_ncbi">Selecting correctly resolved names in <code>TNRS_NCBI</code></h2>
<h3 id="family-level-2">Family level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.family &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;family&quot;</span> &amp;
<span class="st">                                </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span>|
<span class="st">                                 </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                </span>tnrs.ncbi$Family_score &gt;<span class="st"> </span><span class="fl">0.85</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.family)</code></pre></div>
<h3 id="genus-level-2">Genus level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.genus<span class="fl">.1</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                                 </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                  </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                 </span>tnrs.ncbi$Genus_score &gt;<span class="st"> </span><span class="fl">0.89</span> &amp;
<span class="st">                                 </span>tnrs.ncbi$Name_score &gt;<span class="st"> </span><span class="fl">0.49</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus<span class="fl">.1</span>)

index.genus<span class="fl">.2</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                                 </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                  </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                 </span>tnrs.ncbi$Genus_score &gt;<span class="st"> </span><span class="fl">0.99</span> &amp;
<span class="st">                                 </span>tnrs.ncbi$Name_score &gt;<span class="st"> </span><span class="fl">0.2</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus<span class="fl">.2</span>)

index.genus<span class="fl">.3</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                                 </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span> &amp;
<span class="st">                                 </span>tnrs.ncbi$Genus_score &gt;<span class="st"> </span><span class="fl">0.88</span> &amp;
<span class="st">                                 </span>tnrs.ncbi$Name_score &gt;<span class="st"> </span><span class="fl">0.49</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus<span class="fl">.3</span>)

index.genus &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.genus<span class="fl">.1</span>, index.genus<span class="fl">.2</span>, index.genus<span class="fl">.3</span>))
<span class="kw">length</span>(index.genus)</code></pre></div>
<h3 id="species-level-2">Species level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.species<span class="fl">.1</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                   </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                    </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                   </span>tnrs.ncbi$Name_score &gt;<span class="st"> </span><span class="fl">0.94</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species<span class="fl">.1</span>)

index.species<span class="fl">.2</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                   </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                    </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                   </span>tnrs.ncbi$Genus_score &gt;<span class="st"> </span><span class="fl">0.81</span> &amp;
<span class="st">                                   </span>tnrs.ncbi$Name_score &gt;<span class="st"> </span><span class="fl">0.51</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species<span class="fl">.2</span>)

index.species<span class="fl">.3</span> &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>(tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                   </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span>  &amp;
<span class="st">                                   </span>tnrs.ncbi$Genus_score &gt;<span class="st"> </span><span class="fl">0.7</span> &amp;
<span class="st">                                   </span>tnrs.ncbi$Specific_epithet_score &gt;<span class="st"> </span><span class="fl">0.75</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species<span class="fl">.3</span>)

index.species &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.species<span class="fl">.1</span>, index.species<span class="fl">.2</span>, index.species<span class="fl">.3</span>))
<span class="kw">length</span>(index.species)</code></pre></div>
<h3 id="variety-level-2">Variety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.var &lt;-<span class="st"> </span>tnrs.ncbi[<span class="kw">which</span>((tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span> |
<span class="st">                              </span>tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;unknown&quot;</span> |
<span class="st">                              </span>tnrs.ncbi$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span>) &amp;
<span class="st">                             </span>(tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                              </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span> |
<span class="st">                              </span>tnrs.ncbi$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.var)

index.ncbi &lt;-<span class="st"> </span><span class="kw">c</span>(index.family, index.genus, index.species, index.var)
<span class="kw">length</span>((index.ncbi))</code></pre></div>
<h3 id="select-certain-or-uncertain-names-1">Select <code>certain</code> or <code>uncertain</code> names</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain &lt;-<span class="st"> </span>tnrs.ncbi[tnrs.ncbi$Name_number %in%<span class="st"> </span>index.ncbi ==<span class="st"> </span>T,]
<span class="kw">dim</span>(tnrs.ncbi.certain)
<span class="kw">save</span>(tnrs.ncbi.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.certain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.ncbi.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.certain.csv&quot;</span>)

tnrs.ncbi.uncertain &lt;-<span class="st"> </span>tnrs.ncbi[tnrs.ncbi$Name_number %in%<span class="st"> </span>index.ncbi ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(tnrs.ncbi.uncertain)
<span class="kw">save</span>(tnrs.ncbi.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.ncbi.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.uncertain.csv&quot;</span>)</code></pre></div>
<h3 id="resolve-remaining-uncertain-names">Resolve remaining <code>uncertain</code> names</h3>
<p><code>tnrs.ncbi</code> still contained 1,464 <code>uncertain</code> names that were resolved in the following way:</p>
<ul>
<li><p>679 names (mainly trivial names) were selected manually and corrected by JD.</p></li>
<li><p>of those 679 names, 62 were corrected using the matching tools on the <a href="link%20in%20here">TPL webpage</a> (e.g. <code>Dicra vagin var. clathrata</code> → <code>Dicranella vaginata</code>).</p></li>
<li><p>the remaining 785 names were checked manually.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">JD<span class="fl">.679</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;JD.679.csv&quot;</span>)
<span class="kw">head</span>(JD<span class="fl">.679</span>)</code></pre></div>
<p>Manually correct, in a spread sheat program, some further mispellings (e.g. <code>ABIESNORD.</code>, <code>xboris</code>). Select <code>non-JD.corrected</code> and correct species in <code>tnrs.ncbi.uncertain.corrected.csv</code>[6].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.ncbi.uncertain.corrected.csv&quot;</span>)

<span class="kw">str</span>(ncbi.uncertain.corr)</code></pre></div>
<p>Select the 679 JD-species, incl. the 62 species that were corrected manually.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.JD.corrected &lt;-<span class="st"> </span>ncbi.uncertain.corr[<span class="kw">which</span>((ncbi.uncertain.corr$for_JD ==
<span class="st">                                                      &quot;x&quot;</span> |
<span class="st">                                                      </span>ncbi.uncertain.corr$corrected !=
<span class="st">                                                      &quot;&quot;</span>)), <span class="dv">2</span>] 
<span class="kw">length</span>(index.JD.corrected)</code></pre></div>
<p><strong>Further select some correct species within genus</strong></p>
<p>Some species were cut down to the genus level by TNRS, although they might be resolvable using the name matching tools on the <a href="http://www.theplantlist.org/">TPL webpage</a>. Select those names for further name matching.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.correct.genus &lt;-<span class="st"> </span>ncbi.uncertain.corr[<span class="kw">which</span>((ncbi.uncertain.corr$Name_matched_rank ==
<span class="st">                                                  &quot;genus&quot;</span> &amp;
<span class="st">                                                  </span>ncbi.uncertain.corr$Taxonomic_status ==
<span class="st">                                                  &quot;Accepted&quot;</span> &amp;
<span class="st">                                                  </span>ncbi.uncertain.corr$Overall_score &gt;<span class="st"> </span><span class="fl">0.6</span>))
                                         , <span class="dv">2</span>] 
<span class="kw">length</span>(index.correct.genus)</code></pre></div>
<p><strong>Further select some correct names that were matched at the rank of <code>species</code></strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.correct.species &lt;-<span class="st"> </span>ncbi.uncertain.corr[<span class="kw">which</span>((ncbi.uncertain.corr$Name_matched_rank ==
<span class="st">                                                    &quot;species&quot;</span> &amp;
<span class="st">                                                    </span>ncbi.uncertain.corr$Taxonomic_status ==
<span class="st">                                                    &quot;Accepted&quot;</span> &amp;
<span class="st">                                                    </span>ncbi.uncertain.corr$Overall_score &gt;<span class="st"> </span><span class="fl">0.89</span>))
                                           , <span class="dv">2</span>] 
<span class="kw">length</span>(index.correct.species)</code></pre></div>
<p><strong>Create an index for those names that could be further corrected:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.ncbi &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.JD.corrected, index.correct.genus, index.correct.species))
<span class="kw">length</span>(index.ncbi)</code></pre></div>
<h3 id="identify-certain-and-uncertain-species">Identify <code>certain</code> and <code>uncertain</code> species</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.certain &lt;-<span class="st"> </span>ncbi.uncertain.corr[ncbi.uncertain.corr$Name_number %in%
<span class="st">                                                   </span>index.ncbi ==<span class="st"> </span>T,]
<span class="kw">dim</span>(ncbi.uncertain.corr.certain)
<span class="kw">save</span>(ncbi.uncertain.corr.certain, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.certain.Rdata&quot;</span>) 
<span class="kw">write.csv</span>(ncbi.uncertain.corr.certain, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.certain.csv&quot;</span>)

ncbi.uncertain.corr.uncertain &lt;-<span class="st"> </span>ncbi.uncertain.corr[ncbi.uncertain.corr$Name_number %in%
<span class="st">                                                     </span>index.ncbi ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(ncbi.uncertain.corr.uncertain)
<span class="kw">save</span>(ncbi.uncertain.corr.uncertain, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(ncbi.uncertain.corr.uncertain, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.uncertain.csv&quot;</span>)</code></pre></div>
<h1 id="using-the-plant-list-matching-tools-for-unresolved-names">Using <code>The Plant List</code> matching tools for unresolved names</h1>
<p>Generate names list from <code>ncbi.uncertain.corr.uncertain</code> to be matched against <code>The Plant List</code>, using <code>Taxonstand::TPL</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain &lt;-<span class="st"> </span><span class="kw">as.character</span>(ncbi.uncertain.corr.uncertain$Name_submitted)</code></pre></div>
<h2 id="lists-of-names-for-tpl">Lists of names for <code>TPL</code></h2>
<p><strong>Run &quot;raw&quot; list against <code>TPL</code>:</strong></p>
<p>Set large edit distance to allow for fuzzy matching, and strip some characters that would prevend matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">TPL</span>(ncbi.uncertain, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">9</span>, <span class="dt">max.distance =</span> <span class="dv">9</span>)
tpl.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;[&quot;</span>, <span class="st">&quot;&quot;</span>, tpl.ncbi<span class="fl">.1</span>)
tpl.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;]&quot;</span>, <span class="st">&quot;&quot;</span>, tpl.ncbi<span class="fl">.1</span>)
tpl.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;|&quot;</span>, <span class="st">&quot;&quot;</span>, tpl.ncbi<span class="fl">.1</span>)
tpl.ncbi<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;?&quot;</span>, <span class="st">&quot;&quot;</span>, tpl.ncbi<span class="fl">.1</span>)
<span class="kw">write.csv</span>(tpl.ncbi<span class="fl">.1</span>, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.1.csv&quot;</span>)</code></pre></div>
<p><strong>Extent each word by <code>*</code>, to allow for even 'fuzzier' matching:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;* &quot;</span>, ncbi.uncertain), <span class="st">&quot;*&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
tpl.ncbi<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">TPL</span>(ncbi.uncertain<span class="fl">.2</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">9</span>, <span class="dt">max.distance =</span> <span class="dv">9</span>)
<span class="kw">write.csv</span>(tpl.ncbi<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.2.csv&quot;</span>)</code></pre></div>
<p><strong>Truncate each word to its first 3 (5 or 7) letters and extent by <code>*</code>, to do more fuzzy matching</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">simpleCap3 &lt;-<span class="st"> </span>function(x) {
  s &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x, <span class="st">&quot; &quot;</span>)[[<span class="dv">1</span>]]
  <span class="kw">paste</span>(<span class="kw">paste</span>(<span class="kw">substring</span>(s, <span class="dv">1</span>,<span class="dv">1</span>), <span class="kw">substring</span>(s, <span class="dv">2</span>,<span class="dv">3</span>), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>, <span class="dt">collapse=</span><span class="st">&quot;* &quot;</span>), <span class="st">&quot;*&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
}

simpleCap5 &lt;-<span class="st"> </span>function(x) {
  s &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x, <span class="st">&quot; &quot;</span>)[[<span class="dv">1</span>]]
  <span class="kw">paste</span>(<span class="kw">paste</span>(<span class="kw">substring</span>(s, <span class="dv">1</span>,<span class="dv">1</span>), <span class="kw">substring</span>(s, <span class="dv">2</span>,<span class="dv">5</span>), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>, <span class="dt">collapse=</span><span class="st">&quot;* &quot;</span>), <span class="st">&quot;*&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
}

simpleCap7 &lt;-<span class="st"> </span>function(x) {
  s &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x, <span class="st">&quot; &quot;</span>)[[<span class="dv">1</span>]]
  <span class="kw">paste</span>(<span class="kw">paste</span>(<span class="kw">substring</span>(s, <span class="dv">1</span>,<span class="dv">1</span>), <span class="kw">substring</span>(s, <span class="dv">2</span>,<span class="dv">7</span>), <span class="dt">sep=</span><span class="st">&quot;&quot;</span>, <span class="dt">collapse=</span><span class="st">&quot;* &quot;</span>), <span class="st">&quot;*&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
}</code></pre></div>
<p>Apply the string truncation functions above:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">sapply</span>(ncbi.uncertain, simpleCap3)</code></pre></div>
<p>Repeat, but instead using <code>simpleCap5</code> or <code>simpleCap7</code>. Further, remove some strings to improve name matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;[&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;]&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;|&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;?&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;+&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;.&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;&lt;&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
ncbi.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;/&quot;</span>, <span class="st">&quot;&quot;</span>, ncbi.uncertain<span class="fl">.3</span>, <span class="dt">fixed =</span> T)
<span class="kw">str</span>(ncbi.uncertain<span class="fl">.3</span>)

tpl.ncbi<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">TPL</span>(ncbi.uncertain<span class="fl">.3</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">9</span>, <span class="dt">max.distance =</span> <span class="dv">9</span>)
<span class="kw">write.csv</span>(tpl.ncbi<span class="fl">.3</span>, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.3.csv&quot;</span>)

tpl.ncbi<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">TPL</span>(ncbi.uncertain<span class="fl">.5</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">9</span>, <span class="dt">max.distance =</span> <span class="dv">9</span>)
<span class="kw">write.csv</span>(tpl.ncbi<span class="fl">.5</span>, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.5.csv&quot;</span>)

tpl.ncbi<span class="fl">.7</span> &lt;-<span class="st"> </span><span class="kw">TPL</span>(ncbi.uncertain<span class="fl">.7</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">9</span>, <span class="dt">max.distance =</span> <span class="dv">9</span>)
<span class="kw">write.csv</span>(tpl.ncbi<span class="fl">.7</span>, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.7.csv&quot;</span>)</code></pre></div>
<p><strong>Combine <code>tpl.ncbi</code> tables:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.ncbi &lt;-<span class="st"> </span><span class="kw">cbind</span>(tpl.ncbi<span class="fl">.1</span>[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)], tpl.ncbi<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)],
                  tpl.ncbi<span class="fl">.3</span>[,<span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)], tpl.ncbi<span class="fl">.5</span>[,<span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)],
                  tpl.ncbi<span class="fl">.7</span>[,<span class="kw">c</span>(<span class="dv">6</span>,<span class="dv">8</span>,<span class="dv">10</span>,<span class="dv">12</span>)])

<span class="kw">rownames</span>(tpl.ncbi) &lt;-<span class="st"> </span><span class="kw">rownames</span>(tpl.ncbi<span class="fl">.7</span>)
tpl.ncbi &lt;-<span class="st"> </span><span class="kw">cbind</span>(ncbi.uncertain.corr.uncertain[,<span class="dv">1</span>:<span class="dv">2</span>], tpl.ncbi)
<span class="kw">str</span>(tpl.ncbi)
tpl.ncbi &lt;-<span class="st"> </span>tpl.ncbi[,-<span class="dv">1</span>]
<span class="kw">write.csv</span>(tpl.ncbi, <span class="dt">file =</span> <span class="st">&quot;tpl.ncbi.csv&quot;</span>)</code></pre></div>
<p>Manually select correct genera and species in <code>tpl.ncbi</code> csv-file and add columns <code>Genus.correct</code> and <code>Species.correct</code>.</p>
<p>Read the manually corrected <code>tpl.ncbi</code> table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.ncbi<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl.ncbi.csv&quot;</span>)
<span class="kw">str</span>(tpl.ncbi<span class="fl">.2</span>)
<span class="kw">names</span>(tpl.ncbi<span class="fl">.2</span>)</code></pre></div>
<p>Select corrected species and concatenate genus and epithet columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.ncbi<span class="fl">.2</span>$name.correct &lt;-<span class="st"> </span><span class="kw">paste</span>(tpl.ncbi<span class="fl">.2</span>$Genus.correct, tpl.ncbi<span class="fl">.2</span>$Species.correct)
index.corr &lt;-<span class="st"> </span>tpl.ncbi<span class="fl">.2</span>[<span class="kw">which</span>(tpl.ncbi<span class="fl">.2</span>$name.correct !=<span class="st"> &quot; &quot;</span>), <span class="dv">2</span>] </code></pre></div>
<p>Merge <code>ncbi.uncertain</code> correspond and <code>tpl.ncbi.2</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.uncertain<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">join</span>(ncbi.uncertain.corr.uncertain,
                                        tpl.ncbi<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">6</span>,<span class="dv">26</span>:<span class="dv">29</span>)], <span class="dt">by =</span> <span class="st">&quot;Name_number&quot;</span>)

<span class="kw">str</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>)
<span class="kw">names</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>)

<span class="kw">write.csv</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.uncertain.2.csv&quot;</span>)
ncbi.uncertain.corr.uncertain<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;ncbi.uncertain.corr.uncertain.2.csv&quot;</span>)</code></pre></div>
<h3 id="tag-names-that-could-not-be-resolved">Tag names that could not be resolved</h3>
<p>If names were not corrected, set <code>Taxonomic.status == &quot;&quot;</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct[
                                    ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct==<span class="st">&quot;&quot;</span>] &lt;-
<span class="st">    </span>ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Taxonomic.status[
                                        ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct ==<span class="st">&quot;&quot;</span>]

<span class="kw">summary</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct)
<span class="kw">str</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct)</code></pre></div>
<p>... and assign <code>No suitable matches found.</code> to the remaining species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct &lt;-
<span class="st">    </span><span class="kw">as.character</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct)
ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct
[<span class="kw">is.na</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct)] &lt;-<span class="st"> &quot;No suitable matches found.&quot;</span></code></pre></div>
<p>Add uncorrected names in column <code>X</code> to <code>name.correct</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$name.correct[
                                    ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Genus.correct==<span class="st">&quot;&quot;</span>] &lt;-
<span class="st">    </span><span class="kw">as.character</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>[,<span class="dv">41</span>])[
        ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Genus.correct==<span class="st">&quot;&quot;</span>]</code></pre></div>
<p>Assign <code>No suitable matches found.</code> to remaining species in <code>name.correct</code> according to <code>Status.correct</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$name.correct[ncbi.uncertain.corr.uncertain<span class="fl">.2</span>$Status.correct==
<span class="st">                                             &quot;No suitable matches found.&quot;</span>] &lt;-
<span class="st">    &quot;No suitable matches found.&quot;</span>

<span class="kw">write.csv</span>(ncbi.uncertain.corr.uncertain<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.uncertain.2.csv&quot;</span>)</code></pre></div>
<p>Done! Use <code>ncbi.uncertain.corr.uncertain.2</code> for later merging with the other data sets.</p>
<h3 id="check-ncbi.certain">Check <code>ncbi.certain</code></h3>
<p>Create extra column <code>species.correct</code> and add the 679 species from <code>status=accepted</code> in <code>JD.corrected</code>. Assign status <code>No suitable matches found.</code> to the remaining non-resolved species.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">JD.correct &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/Analyses/</span>
<span class="st">Code/Unresolved_sPlot2.0_TRY3.0_2_JD.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</code></pre></div>
<h2 id="assign-no-suitable-matches-found.-to-non-correctable-species">Assign <code>No suitable matches found.</code> to non-correctable species</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">JD.correct$Taxon[JD.correct$Taxon==<span class="st">&quot;&quot;</span>] &lt;-<span class="st"> &quot;No suitable matches found.&quot;</span>
<span class="kw">str</span>(JD.correct)
<span class="kw">str</span>(ncbi.uncertain.corr.certain)</code></pre></div>
<p>Join to <code>ncbi.uncertain.corr.certain</code> based on <code>Name_number</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr &lt;-<span class="st"> </span><span class="kw">join</span>(ncbi.uncertain.corr.certain, JD.correct[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">8</span>)],
                                  <span class="dt">by =</span> <span class="st">&quot;Name_number&quot;</span>)
<span class="kw">str</span>(ncbi.certain.JD.corr)
<span class="kw">write.csv</span>(ncbi.certain.JD.corr, <span class="dt">file =</span> <span class="st">&quot;ncbi.certain.JD.corr.csv&quot;</span>)
ncbi.certain.JD.corr<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;ncbi.certain.JD.corr.csv&quot;</span>,
                                        <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(ncbi.certain.JD.corr<span class="fl">.2</span>)</code></pre></div>
<p>Add the corrected names from <code>ncbi.certain.JD.corr.2$corrected</code> to <code>name.correct</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct[<span class="kw">is.na</span>(ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct)] &lt;-
<span class="st">    </span>ncbi.certain.JD.corr<span class="fl">.2</span>$corrected[<span class="kw">is.na</span>(ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct)]
<span class="kw">write.csv</span>(ncbi.certain.JD.corr<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.certain.JD.corr.2.csv&quot;</span>)</code></pre></div>
<p>Fill the missing names in <code>name.correct</code>, because they were correctly resolved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct[ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct==<span class="st">&quot;&quot;</span>] &lt;-
<span class="st">    </span>ncbi.certain.JD.corr<span class="fl">.2</span>$Name_matched[ncbi.certain.JD.corr<span class="fl">.2</span>$name.correct==<span class="st">&quot;&quot;</span>]</code></pre></div>
<p>All entries in <code>ncbi.certain.JD.corr.2</code> are assigned to a name. Now merge it with <code>ncbi.uncertain.corr.uncertain.2</code> (~1,400 names). Add tag <code>manual matching</code>. Means the scores from TNRS matching are useless here, but nevertheless keep them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;ncbi.certain.JD.corr.3.csv&quot;</span>,
                                   <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(ncbi.certain.JD.corr<span class="fl">.3</span>)
ncbi.uncertain.corr.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;ncbi.uncertain.corr.uncertain.3.csv&quot;</span>,
                                          <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(ncbi.uncertain.corr.uncertain<span class="fl">.3</span>)</code></pre></div>
<p>Add status <code>No suitable matches found.</code> or <code>Accepted</code> in <code>ncbi.certain.JD.corr.3</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr<span class="fl">.3</span>$Status.correct[ncbi.certain.JD.corr<span class="fl">.3</span>$name.correct ==
<span class="st">                                           &quot;No suitable matches found.&quot;</span>] &lt;-
<span class="st">    &quot;No suitable matches found.&quot;</span>
ncbi.certain.JD.corr<span class="fl">.3</span>$Status.correct[
                                <span class="kw">is.na</span>(ncbi.certain.JD.corr<span class="fl">.3</span>$Status.correct)] &lt;-
<span class="st">    &quot;Accepted&quot;</span>
<span class="kw">write.csv</span>(ncbi.certain.JD.corr<span class="fl">.3</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.certain.JD.corr.3.csv&quot;</span>)</code></pre></div>
<p>In <code>ncbi.certain.JD.corr.3</code>, assign all entries an <code>x</code> for <code>Manual matching</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.certain.JD.corr<span class="fl">.3</span>$Manual.matching &lt;-<span class="st"> &quot;x&quot;</span>
ncbi.uncertain.corr.uncertain<span class="fl">.3</span>$Manual.matching &lt;-<span class="st"> &quot;x&quot;</span>
<span class="kw">write.csv</span>(ncbi.certain.JD.corr<span class="fl">.3</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.certain.JD.corr.3.csv&quot;</span>)
<span class="kw">write.csv</span>(ncbi.uncertain.corr.uncertain<span class="fl">.3</span>, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.corr.uncertain.3.csv&quot;</span>)</code></pre></div>
<h3 id="combine-the-uncertain-and-certain-species-lists">Combine the <code>uncertain</code> and <code>certain</code> species lists</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(ncbi.certain.JD.corr<span class="fl">.3</span>)
<span class="kw">names</span>(ncbi.uncertain.corr.uncertain<span class="fl">.3</span>)

ncbi.certain.JD.corr<span class="fl">.3</span> &lt;-<span class="st"> </span>ncbi.certain.JD.corr<span class="fl">.3</span>[,-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>)]
ncbi.uncertain.corr.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span>ncbi.uncertain.corr.uncertain<span class="fl">.3</span>[,-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)]

<span class="kw">match</span>(<span class="kw">names</span>(ncbi.uncertain.corr.uncertain<span class="fl">.3</span>), <span class="kw">names</span>(ncbi.certain.JD.corr<span class="fl">.3</span>))

ncbi.uncertain.comb &lt;-<span class="st"> </span><span class="kw">rbind</span>(ncbi.uncertain.corr.uncertain<span class="fl">.3</span>, ncbi.certain.JD.corr<span class="fl">.3</span>)
<span class="kw">dim</span>(ncbi.uncertain.comb)
<span class="kw">write.csv</span>(ncbi.uncertain.comb, <span class="dt">file =</span> <span class="st">&quot;ncbi.uncertain.comb.csv&quot;</span>)</code></pre></div>
<p>Read in <code>tnrs.ncbi.certain</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.ncbi.certain.csv&quot;</span>,
                              <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(tnrs.ncbi.certain)</code></pre></div>
<h2 id="resolve-names-that-were-reduced-to-genus-level-by-tnrs">Resolve names that were reduced to genus-level by TNRS</h2>
<p>Read in <code>tnrs.ncbi.certain&quot;</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.ncbi.certain.csv&quot;</span>,
                              <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(tnrs.ncbi.certain)</code></pre></div>
<p>To resolve names that were reduced to genus-level, identify <code>Name_submitted</code> where <code>Overall_score</code> == 0.5:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index0<span class="fl">.5</span> &lt;-<span class="st"> </span>tnrs.ncbi.certain[<span class="kw">which</span>(tnrs.ncbi.certain$Overall_score ==<span class="st"> </span><span class="fl">0.5</span>), <span class="dv">2</span>]
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span>tnrs.ncbi.certain$Name_submitted[tnrs.ncbi.certain$Overall_score ==<span class="st"> </span><span class="fl">0.5</span>]
<span class="kw">names</span>(tpl0<span class="fl">.5</span>) &lt;-<span class="st"> </span>index0<span class="fl">.5</span>
<span class="kw">str</span>(tpl0<span class="fl">.5</span>)
<span class="kw">length</span>(tpl0<span class="fl">.5</span>)
tpl0<span class="fl">.5</span>[<span class="dv">1</span>:<span class="dv">100</span>]
<span class="kw">save</span>(tpl0<span class="fl">.5</span>, <span class="dt">file =</span> <span class="st">&quot;tpl0.5.Rdata&quot;</span>)</code></pre></div>
<p>Do some string cleaning to improve matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;[&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;]&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;|&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;?&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;+&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;.&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;&lt;&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)
tpl0<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;/&quot;</span>, <span class="st">&quot;&quot;</span>, tpl0<span class="fl">.5</span>, <span class="dt">fixed =</span> T)

<span class="kw">system.time</span>(
tpl0<span class="fl">.5</span>.res &lt;-<span class="st"> </span><span class="kw">TPL</span>(tpl0<span class="fl">.5</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">2</span>, <span class="dt">max.distance =</span> <span class="dv">1</span>)
)</code></pre></div>
<p>Extent each word by <code>*</code>, to allow for even 'fuzzier' matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.csv</span>(tpl0<span class="fl">.5</span>.res, <span class="dt">file =</span> <span class="st">&quot;tpl0.5.res.csv&quot;</span>)
tpl0<span class="fl">.5.2</span> &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;* &quot;</span>, tpl0<span class="fl">.5</span>), <span class="st">&quot;*&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
tpl0<span class="fl">.5.2</span>.res &lt;-<span class="st"> </span><span class="kw">TPL</span>(tpl0<span class="fl">.5.2</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">2</span>, <span class="dt">max.distance =</span> <span class="dv">1</span>)
<span class="kw">write.csv</span>(tpl0<span class="fl">.5.2</span>.res, <span class="dt">file =</span> <span class="st">&quot;tpl0.5.2.res.csv&quot;</span>)</code></pre></div>
<p>Truncate each word to its first 5 (or 7) letters and extent words by <code>*</code>, to do more fuzzy matching:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl0<span class="fl">.5.5</span> &lt;-<span class="st"> </span><span class="kw">sapply</span>(tpl0<span class="fl">.5</span>, simpleCap5)
tpl0<span class="fl">.5.5</span>.res &lt;-<span class="st"> </span><span class="kw">TPL</span>(tpl0<span class="fl">.5.5</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">2</span>, <span class="dt">max.distance =</span> <span class="dv">1</span>)
<span class="kw">write.csv</span>(tpl0<span class="fl">.5.5</span>.res, <span class="dt">file =</span> <span class="st">&quot;tpl0.5.5.res.csv&quot;</span>)

tpl0<span class="fl">.5.7</span> &lt;-<span class="st"> </span><span class="kw">sapply</span>(tpl0<span class="fl">.5</span>, simpleCap7)
tpl0<span class="fl">.5.7</span>.res &lt;-<span class="st"> </span><span class="kw">TPL</span>(tpl0<span class="fl">.5.7</span>, <span class="dt">corr=</span>T, <span class="dt">diffchar =</span> <span class="dv">2</span>, <span class="dt">max.distance =</span> <span class="dv">1</span>)
<span class="kw">write.csv</span>(tpl0<span class="fl">.5.7</span>.res, <span class="dt">file =</span> <span class="st">&quot;tpl0.5.7.res.csv&quot;</span>)

tpl0<span class="fl">.5</span>.res &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl0.5.res.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
tpl0<span class="fl">.5.2</span>.res &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl0.5.2.res.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
tpl0<span class="fl">.5.5</span>.res &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl0.5.5.res.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
tpl0<span class="fl">.5.7</span>.res &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl0.5.7.res.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>Combine <code>tpl.ncbi</code> tables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.res.comb &lt;-<span class="st"> </span><span class="kw">cbind</span>(tpl0<span class="fl">.5</span>.res[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>)],
tpl0<span class="fl">.5.2</span>.res[,<span class="kw">c</span>(<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>)], tpl0<span class="fl">.5.5</span>.res[,<span class="kw">c</span>(<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>)],
tpl0<span class="fl">.5.7</span>.res[,<span class="kw">c</span>(<span class="dv">7</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>)])
                                                                                   
<span class="kw">head</span>(tpl.res.comb)
<span class="kw">write.csv</span>(tpl.res.comb, <span class="dt">file =</span> <span class="st">&quot;tpl.res.comb.csv&quot;</span>)
tpl.res.comb &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl.res.comb.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</code></pre></div>
<h3 id="manually-add-and-fill-in-new-columns-to-tpl.res.comb.csv">Manually add and fill in new columns to <code>tpl.res.comb.csv</code>:</h3>
<ul>
<li><code>Status.correct</code></li>
<li><code>Genus.correct</code></li>
<li><code>Species.correct</code></li>
</ul>
<h3 id="combine-species.correct-and-genus.correct">Combine <code>species.correct</code> and <code>genus.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tpl.res.comb &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tpl.res.comb.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(tpl.res.comb)
<span class="kw">names</span>(tpl.res.comb)
tpl.res.comb$name.correct &lt;-<span class="st"> </span><span class="kw">paste</span>(tpl.res.comb$Genus.correct,
tpl.res.comb$Species.correct)
tpl.res.comb<span class="fl">.2</span> &lt;-<span class="st"> </span>tpl.res.comb</code></pre></div>
<h3 id="join-tnrs.ncbi.certain-and-tpl.res.comb.2">Join <code>tnrs.ncbi.certain</code> and <code>tpl.res.comb.2</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(tpl.res.comb<span class="fl">.2</span>)
<span class="kw">names</span>(tpl.res.comb<span class="fl">.2</span>)[<span class="dv">2</span>] &lt;-<span class="st"> &quot;Name_number&quot;</span>
<span class="kw">names</span>(tnrs.ncbi.certain)</code></pre></div>
<h3 id="get-match-correct">Get match correct</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain<span class="fl">.0.5</span> &lt;-<span class="st"> </span><span class="kw">join</span>(tnrs.ncbi.certain[<span class="kw">match</span>(index0<span class="fl">.5</span>,
                                                      tnrs.ncbi.certain$Name_number),],
                              tpl.res.comb<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">23</span>:<span class="dv">26</span>)], <span class="dt">by =</span>
<span class="st">&quot;Name_number&quot;</span>)
<span class="kw">write.csv</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.certain.0.5.csv&quot;</span>)</code></pre></div>
<p>Fill in extra columns in <code>tnrs.ncbi.certain.0.5</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain<span class="fl">.0.5</span> &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tnrs.ncbi.certain.0.5.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>)
<span class="kw">names</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>)
<span class="kw">str</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>$Genus.correct)</code></pre></div>
<p>Fill <code>Manual.matching</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain<span class="fl">.0.5</span>$Manual.matching[tnrs.ncbi.certain<span class="fl">.0.5</span>$Genus.correct !=<span class="st"> &quot;&quot;</span>] &lt;-<span class="st"> &quot;x&quot;</span></code></pre></div>
<p>Fill <code>Status.correct</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain<span class="fl">.0.5</span>$Status.correct[tnrs.ncbi.certain<span class="fl">.0.5</span>$Status.correct ==<span class="st"> &quot;&quot;</span>] &lt;-
<span class="st">    </span>tnrs.ncbi.certain<span class="fl">.0.5</span>$Taxonomic_status[tnrs.ncbi.certain<span class="fl">.0.5</span>$Status.correct ==<span class="st"> &quot;&quot;</span>]</code></pre></div>
<p>Fill <code>name.correct</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain<span class="fl">.0.5</span>$name.correct[tnrs.ncbi.certain<span class="fl">.0.5</span>$name.correct ==<span class="st"> &quot; &quot;</span>] &lt;-
<span class="st">    </span>tnrs.ncbi.certain<span class="fl">.0.5</span>$Name_matched[tnrs.ncbi.certain<span class="fl">.0.5</span>$name.correct ==<span class="st"> &quot; &quot;</span>]

<span class="kw">write.csv</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.certain.0.5.csv&quot;</span>)</code></pre></div>
<p>Combine <code>tnrs.ncbi.certain.0.5</code> with the remaining <code>tnrs.ncbi.certain</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>)
<span class="kw">names</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>)

cert<span class="fl">.0.5</span> &lt;-<span class="st"> </span>tnrs.ncbi.certain[tnrs.ncbi.certain$Overall_score ==<span class="st"> </span><span class="fl">0.5</span>,]
<span class="kw">dim</span>(cert<span class="fl">.0.5</span>)
<span class="kw">str</span>(cert<span class="fl">.0.5</span>)
cert.non<span class="fl">.0.5</span> &lt;-<span class="st"> </span>tnrs.ncbi.certain[tnrs.ncbi.certain$Overall_score !=<span class="st"> </span><span class="fl">0.5</span>,]
<span class="kw">dim</span>(cert.non<span class="fl">.0.5</span>)
<span class="kw">str</span>(cert.non<span class="fl">.0.5</span>)
<span class="kw">names</span>(cert.non<span class="fl">.0.5</span>)</code></pre></div>
<p>Add three more columns to <code>cert.non.0.5</code> so that the columns to those in <code>tnrs.ncbi.certain.0.5</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cert.non<span class="fl">.0.5</span>$Manual.matching &lt;-<span class="st"> </span><span class="ot">NA</span>
cert.non<span class="fl">.0.5</span>$Status.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
cert.non<span class="fl">.0.5</span>$name.correct &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<p>Combine the two tables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.ncbi.certain.comb &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.ncbi.certain<span class="fl">.0.5</span>[,<span class="kw">c</span>(<span class="dv">3</span>:<span class="dv">41</span>,<span class="dv">44</span>)], cert.non<span class="fl">.0.5</span>[,<span class="kw">c</span>(<span class="dv">2</span>:<span class="dv">41</span>)])
<span class="kw">dim</span>(tnrs.ncbi.certain.comb)
<span class="kw">write.csv</span>(tnrs.ncbi.certain.comb, <span class="dt">file =</span> <span class="st">&quot;tnrs.ncbi.certain.comb.csv&quot;</span>)</code></pre></div>
<h1 id="merge-the-resolved-species-lists">Merge the resolved species lists</h1>
<h2 id="read-files">Read files</h2>
<h3 id="tpl.small.certain"><code>TPL.small.certain</code></h3>
<p>Contains TNRS results based on the five sources: TPL, GCC, ILDIS, Tropicos and USDA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;tnrs.tpl.certain.Rdata&quot;</span>)
<span class="kw">dim</span>(tnrs.tpl.certain)</code></pre></div>
<h3 id="trop.small.certain"><code>Trop.small.certain</code></h3>
<p>Contains TNRS results based on the five sources Tropicos (ranked first) TPL, GCC, ILDIS &amp; USDA</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;tnrs.trop.small.certain.Rdata&quot;</span>)
<span class="kw">dim</span>(tnrs.trop.small.certain)</code></pre></div>
<p>Combine the <code>certain</code> data sets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.certain &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.tpl.certain, tnrs.trop.small.certain)
<span class="kw">dim</span>(tnrs.tpl.all.trop.certain)</code></pre></div>
<p>... and add the four additional columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(tnrs.tpl.all.trop.certain)

tnrs.tpl.all.trop.certain$Manual.matching &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.tpl.all.trop.certain$Status.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.tpl.all.trop.certain$name.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.tpl.all.trop.certain$rank.correct &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h3 id="pick-the-respective-ncbi-data-sets">Pick the respective <code>NCBI</code> data sets</h3>
<p>... for the 8,177 certain species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(tnrs.ncbi.certain.comb)
tnrs.ncbi.certain.comb$rank.correct &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<p>Combine the with the big list above:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.certain<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.tpl.all.trop.certain, tnrs.ncbi.certain.comb)
<span class="kw">dim</span>(tnrs.tpl.all.trop.certain<span class="fl">.2</span>)
<span class="kw">names</span>(tnrs.tpl.all.trop.certain<span class="fl">.2</span>)</code></pre></div>
<p>Pick the list containing the <code>uncertain</code> species:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(ncbi.uncertain.comb)</code></pre></div>
<p>Exclude columns <code>JD</code> and <code>corrected</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ncbi.uncertain.comb<span class="fl">.2</span> &lt;-<span class="st"> </span>ncbi.uncertain.comb[,-<span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">6</span>)]
<span class="kw">names</span>(ncbi.uncertain.comb<span class="fl">.2</span>)
ncbi.uncertain.comb<span class="fl">.2</span>$rank.correct &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<p>Combine them big list containing the <code>certain</code> species with the <code>uncertain</code> species</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.tpl.all.trop.certain<span class="fl">.2</span>, ncbi.uncertain.comb<span class="fl">.2</span>)
<span class="kw">write.csv</span>(tnrs.tpl.all.trop.tnrs.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.csv&quot;</span>)
<span class="kw">save</span>(tnrs.tpl.all.trop.tnrs.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.Rdata&quot;</span>)</code></pre></div>
<p>Correct one more species name:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain$name.correct[<span class="kw">which</span>(tnrs.tpl.all.trop.tnrs.certain$name.correct
                                                  ==<span class="st"> &quot;ABIES NORDMANNIANA&quot;</span>)] &lt;-
<span class="st">    &quot;Abies nordmanniana&quot;</span></code></pre></div>
<h2 id="complement-the-main-columns-in-the-backbone">Complement the main columns in the backbone</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(tnrs.tpl.all.trop.tnrs.certain)
<span class="kw">names</span>(tnrs.tpl.all.trop.tnrs.certain)</code></pre></div>
<h3 id="fill-in-rank.correct">Fill in <code>rank.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled &lt;-<span class="st"> </span>tnrs.tpl.all.trop.tnrs.certain

tnrs.tpl.all.trop.tnrs.certain.filled$rank.correct &lt;-<span class="st"> </span><span class="kw">as.character</span>(
    tnrs.tpl.all.trop.tnrs.certain.filled$rank.correct)

tnrs.tpl.all.trop.tnrs.certain.filled$rank.correct[
                                          <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                </span>Manual.matching)] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$Name_matched_rank[
                                              <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                    </span>Manual.matching)]</code></pre></div>
<h3 id="fill-status.correct">Fill <code>status.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$Status.correct[
                                          <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                </span>Manual.matching)] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$Taxonomic_status[
                                              <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                    </span>Manual.matching)]</code></pre></div>
<h3 id="fill-name.correct">Fill <code>name.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[<span class="kw">is.na</span>(
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>Manual.matching)] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$Accepted_name[<span class="kw">is.na</span>(
                                              tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                              </span>Manual.matching)]</code></pre></div>
<p>Fill <code>name.correct</code> where <code>status.correct</code> == <code>No opinion</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[<span class="kw">which</span>(
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>Status.correct ==<span class="st"> &quot;No opinion&quot;</span>)] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$Name_matched[<span class="kw">which</span>(
                                              tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                              </span>Status.correct ==<span class="st"> &quot;No opinion&quot;</span>)]

<span class="kw">write.csv</span>(tnrs.tpl.all.trop.tnrs.certain.filled,
          <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>)</code></pre></div>
<p>In <code>tnrs.tpl.all.trop.tnrs.certain.filled</code> resolve some species manually on the TPL webpage.</p>
<h3 id="fill-in-names.short-correct">Fill in <code>names.short correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>,
                                                  <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">names</span>(tnrs.tpl.all.trop.tnrs.certain.filled)
<span class="kw">str</span>(tnrs.tpl.all.trop.tnrs.certain.filled)
tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct &lt;-<span class="st"> </span><span class="kw">as.character</span>(
    tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct)</code></pre></div>
<h3 id="shorten-names-that-have-more-than-two-words-and-where-the-second-word-is-a-x">Shorten names that have more than two words and where the second word is a <code>x</code></h3>
<p>Because some names consisted of &gt; 2 words, the number of words in each name was counted:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wordcount &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">gregexpr</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">S+&quot;</span>, tnrs.tpl.all.trop.tnrs.certain.filled$name.correct),
                    length)</code></pre></div>
<p>If a name has &gt; 1 word, just keep the first two words:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct[wordcount&gt;<span class="dv">1</span>] &lt;-
<span class="st">    </span><span class="kw">word</span>(tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[wordcount&gt;<span class="dv">1</span>], <span class="dv">1</span>, <span class="dv">2</span>)</code></pre></div>
<p>Fill in one-word names <code>name.short.correct</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct[wordcount==<span class="dv">1</span>] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[wordcount==<span class="dv">1</span>]</code></pre></div>
<p>Select names where the second word is a <code>x</code> (hybrids):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">word</span>(tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct[wordcount&gt;<span class="dv">1</span>], <span class="dv">2</span>)==<span class="st">&quot;x&quot;</span>)

<span class="kw">write.csv</span>(tnrs.tpl.all.trop.tnrs.certain.filled,
          <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>)</code></pre></div>
<p>There are 7,900 species with more than 2 words and where 2nd word is &quot;x&quot;. For those names in <code>name.short.correct</code> (i) where the second word is a <code>x</code> and that (ii) had more than two words, pick the first and third word:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index &lt;-<span class="st"> </span><span class="kw">word</span>(tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[wordcount&gt;<span class="dv">2</span>], <span class="dv">2</span>) ==<span class="st"> &quot;x&quot;</span>
wo.x &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">word</span>(tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[wordcount&gt;<span class="dv">2</span>][index],
<span class="dv">1</span>), <span class="kw">word</span>(tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[wordcount&gt;<span class="dv">2</span>][index], <span class="dv">3</span>))

tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct[wordcount&gt;<span class="dv">2</span>][index] &lt;-<span class="st"> </span>wo.x</code></pre></div>
<h3 id="in-name.short-correct-insert-na-where-no-suitable-matches-found.">In <code>name.short correct</code> insert <code>NA</code> where <code>&quot;No suitable matches found.&quot;</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>,
                                                  <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
tnrs.tpl.all.trop.tnrs.certain.filled$name.short.correct[tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                         </span>name.correct ==
<span class="st">                                                         &quot;No suitable matches found.&quot;</span>] &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h3 id="fill-in-rank.short.correct">Fill in <code>rank.short.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$rank.correct
<span class="kw">table</span>(tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct)</code></pre></div>
<p>Assign <code>rank</code> = <code>species</code> to the shortened names of <code>subspecies</code>, <code>subvariety</code>, <code>variety</code> and <code>forma</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                                         </span>rank.correct ==<span class="st"> &quot;infraspecies&quot;</span>] &lt;-
<span class="st">    &quot;species&quot;</span>

tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>rank.correct ==<span class="st"> &quot;subspecies&quot;</span>] &lt;-<span class="st"> &quot;species&quot;</span>

tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>rank.correct ==<span class="st"> &quot;subvariety&quot;</span>] &lt;-<span class="st"> &quot;species&quot;</span>

tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>rank.correct ==<span class="st"> &quot;variety&quot;</span>] &lt;-<span class="st"> &quot;species&quot;</span>

tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>rank.correct ==<span class="st"> &quot;forma&quot;</span>] &lt;-<span class="st"> &quot;species&quot;</span></code></pre></div>
<p>Assign <code>rank.short.correct</code> = <code>family</code> to <code>rank.correct</code> == <code>unknown</code> since all of those names were at the family level.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct[
                                          tnrs.tpl.all.trop.tnrs.certain.filled$
<span class="st">                                          </span>rank.correct ==<span class="st"> &quot;unknown&quot;</span>] &lt;-<span class="st"> &quot;family&quot;</span>

<span class="kw">write.csv</span>(tnrs.tpl.all.trop.tnrs.certain.filled,
          <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>)</code></pre></div>
<h2 id="fix-families">Fix families</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>,
                                                  <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(tnrs.tpl.all.trop.tnrs.certain.filled)</code></pre></div>
<pre><code>##  [1] &quot;X.3&quot;                             &quot;X.2&quot;                            
##  [3] &quot;X.1&quot;                             &quot;X&quot;                              
##  [5] &quot;Name_number&quot;                     &quot;Name_submitted&quot;                 
##  [7] &quot;Overall_score&quot;                   &quot;Name_matched&quot;                   
##  [9] &quot;Name_matched_rank&quot;               &quot;Name_score&quot;                     
## [11] &quot;Name_matched_author&quot;             &quot;Name_matched_url&quot;               
## [13] &quot;Author_matched&quot;                  &quot;Author_score&quot;                   
## [15] &quot;Family_matched&quot;                  &quot;Family_score&quot;                   
## [17] &quot;Name_matched_accepted_family&quot;    &quot;Genus_matched&quot;                  
## [19] &quot;Genus_score&quot;                     &quot;Specific_epithet_matched&quot;       
## [21] &quot;Specific_epithet_score&quot;          &quot;Infraspecific_rank&quot;             
## [23] &quot;Infraspecific_epithet_matched&quot;   &quot;Infraspecific_epithet_score&quot;    
## [25] &quot;Infraspecific_rank_2&quot;            &quot;Infraspecific_epithet_2_matched&quot;
## [27] &quot;Infraspecific_epithet_2_score&quot;   &quot;Annotations&quot;                    
## [29] &quot;Unmatched_terms&quot;                 &quot;Taxonomic_status&quot;               
## [31] &quot;Accepted_name&quot;                   &quot;Accepted_name_author&quot;           
## [33] &quot;Accepted_name_rank&quot;              &quot;Accepted_name_url&quot;              
## [35] &quot;Accepted_name_species&quot;           &quot;Accepted_name_family&quot;           
## [37] &quot;Selected&quot;                        &quot;Source&quot;                         
## [39] &quot;Warnings&quot;                        &quot;Accepted_name_lsid&quot;             
## [41] &quot;user_id&quot;                         &quot;Manual.matching&quot;                
## [43] &quot;Status.correct&quot;                  &quot;name.correct&quot;                   
## [45] &quot;rank.correct&quot;                    &quot;family.correct&quot;                 
## [47] &quot;name.short.correct&quot;              &quot;rank.short.correct&quot;</code></pre>
<p>Fill in family names in <code>family.correct</code> for entries where the matched rank was <code>family</code> and that were not matched manually:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.all.trop.tnrs.certain.filled$family.correct &lt;-<span class="st"> </span><span class="kw">as.character</span>(
    tnrs.tpl.all.trop.tnrs.certain.filled$family.correct)

tnrs.tpl.all.trop.tnrs.certain.filled$family.correct[(
    <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$Manual.matching) &amp;
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct ==<span class="st"> &quot;family&quot;</span>)] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$name.correct[(
        <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$Manual.matching) &amp;
<span class="st">        </span>tnrs.tpl.all.trop.tnrs.certain.filled$rank.short.correct ==<span class="st"> &quot;family&quot;</span>)]

tnrs.tpl.all.trop.tnrs.certain.filled$family.correct[(
    <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$Manual.matching))] &lt;-
<span class="st">    </span>tnrs.tpl.all.trop.tnrs.certain.filled$Name_matched_accepted_family[(
    <span class="kw">is.na</span>(tnrs.tpl.all.trop.tnrs.certain.filled$Manual.matching))]

<span class="kw">write.csv</span>(tnrs.tpl.all.trop.tnrs.certain.filled,
          <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.all.trop.tnrs.certain.filled.csv&quot;</span>)</code></pre></div>
<h2 id="link-backbone-to-original-taxa-names-and-splottry-code">Link backbone to original taxa names and <code>sPlot/TRY</code> code</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splot.try3.code &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;spec.list.TRY3.sPlot2.csv&quot;</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
<span class="kw">str</span>(splot.try3.code)

backbone.splot.try3 &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../tnrs.tpl.all.trop.tnrs.certain.filled.small.csv&quot;</span>,
                                <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(backbone.splot.try3)</code></pre></div>
<pre><code>## 'data.frame':    122901 obs. of  30 variables:
##  $ Name_number                 : int  1 2 3 4 5 6 7 8 9 10 ...
##  $ Name_submitted              : chr  &quot;Spermatophyta sp.&quot; &quot;Spermatophyta sp.&quot; &quot;Chlorophytum sp. [1269]&quot; &quot;Echinochloa sp.&quot; ...
##  $ Overall_score               : num  0 0 0.9 0.9 0.9 0.9 0.9 0 0.9 0 ...
##  $ Name_matched                : chr  &quot;No suitable matches found.&quot; &quot;No suitable matches found.&quot; &quot;Chlorophytum&quot; &quot;Echinochloa&quot; ...
##  $ Name_matched_rank           : chr  &quot;&quot; &quot;&quot; &quot;genus&quot; &quot;genus&quot; ...
##  $ Name_score                  : num  0 0 1 1 1 1 1 0 1 0 ...
##  $ Family_score                : num  0 0 NA NA NA NA 1 0 1 0 ...
##  $ Name_matched_accepted_family: chr  &quot;&quot; &quot;&quot; &quot;Asparagaceae&quot; &quot;Poaceae&quot; ...
##  $ Genus_matched               : chr  &quot;&quot; &quot;&quot; &quot;Chlorophytum&quot; &quot;Echinochloa&quot; ...
##  $ Genus_score                 : num  0 0 1 1 1 1 NA 0 NA 0 ...
##  $ Specific_epithet_matched    : chr  &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
##  $ Specific_epithet_score      : num  0 0 NA NA NA NA NA 0 NA 0 ...
##  $ Unmatched_terms             : chr  &quot;&quot; &quot;&quot; &quot;\&quot;\&quot;sp. [1269]&quot; &quot;\&quot;\&quot;sp.&quot; ...
##  $ Taxonomic_status            : chr  &quot;&quot; &quot;&quot; &quot;Accepted&quot; &quot;Accepted&quot; ...
##  $ Accepted_name               : chr  &quot;&quot; &quot;&quot; &quot;Chlorophytum&quot; &quot;Echinochloa&quot; ...
##  $ Accepted_name_author        : chr  &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
##  $ Accepted_name_rank          : chr  &quot;&quot; &quot;&quot; &quot;genus&quot; &quot;genus&quot; ...
##  $ Accepted_name_url           : chr  &quot;&quot; &quot;&quot; &quot;http://www.theplantlist.org/tpl1.1/search?q=Chlorophytum&quot; &quot;http://www.theplantlist.org/tpl1.1/search?q=Echinochloa&quot; ...
##  $ Accepted_name_species       : chr  &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
##  $ Accepted_name_family        : chr  &quot;&quot; &quot;&quot; &quot;Asparagaceae&quot; &quot;Poaceae&quot; ...
##  $ Selected                    : chr  &quot;true&quot; &quot;true&quot; &quot;true&quot; &quot;true&quot; ...
##  $ Source                      : chr  &quot;&quot; &quot;&quot; &quot;tpl&quot; &quot;tpl&quot; ...
##  $ Warnings                    : chr  &quot; &quot; &quot; &quot; &quot; &quot; &quot; &quot; ...
##  $ Manual.matching             : chr  NA NA NA NA ...
##  $ Status.correct              : chr  &quot;No suitable matches found.&quot; &quot;No suitable matches found.&quot; &quot;Accepted&quot; &quot;Accepted&quot; ...
##  $ name.correct                : chr  &quot;No suitable matches found.&quot; &quot;No suitable matches found.&quot; &quot;Chlorophytum&quot; &quot;Echinochloa&quot; ...
##  $ rank.correct                : chr  &quot;higher&quot; &quot;higher&quot; &quot;genus&quot; &quot;genus&quot; ...
##  $ family.correct              : chr  &quot;&quot; &quot;&quot; &quot;Asparagaceae&quot; &quot;Poaceae&quot; ...
##  $ name.short.correct          : chr  NA NA &quot;Chlorophytum&quot; &quot;Echinochloa&quot; ...
##  $ rank.short.correct          : chr  &quot;higher&quot; &quot;higher&quot; &quot;genus&quot; &quot;genus&quot; ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.splot.try3 &lt;-<span class="st"> </span><span class="kw">join</span>(splot.try3.code, backbone.splot.try3, <span class="dt">by =</span> <span class="st">&quot;Name_number&quot;</span>)
<span class="kw">str</span>(backbone.splot.try3)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(backbone.splot.try3$Status.correct)</code></pre></div>
<pre><code>## 
##                   Accepted                 No opinion 
##                      95485                       4591 
## No suitable matches found.                    Synonym 
##                       1692                      20952 
##                 Unresolved 
##                        181</code></pre>
<p>For constistency, assign <code>Unresolved</code> to <code>status = Unresolved</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.splot.try3$Status.correct[backbone.splot.try3$Status.correct ==<span class="st"> &quot;No opinion&quot;</span>] &lt;-
<span class="st">    &quot;Unresolved&quot;</span>

<span class="kw">write.csv</span>(backbone.splot.try3, <span class="dt">file =</span> <span class="st">&quot;backbone.splot.try3.csv&quot;</span>)
<span class="kw">save</span>(backbone.splot.try3, <span class="dt">file =</span> <span class="st">&quot;backbone.splot.try3.Rdata&quot;</span>)</code></pre></div>
<h2 id="some-statistics">Some statistics</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(backbone.splot.try3$sPlot.TRY)</code></pre></div>
<pre><code>## &lt; table of extent 0 &gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(backbone.splot.try3$Manual.matching)</code></pre></div>
<pre><code>## 
##    x 
## 1675</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(backbone.splot.try3$Status.correct)</code></pre></div>
<pre><code>## 
##                   Accepted                 No opinion 
##                      95485                       4591 
## No suitable matches found.                    Synonym 
##                       1692                      20952 
##                 Unresolved 
##                        181</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(backbone.splot.try3$name.correct))-<span class="dv">1</span></code></pre></div>
<pre><code>## [1] 90696</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(backbone.splot.try3$family.correct))-<span class="dv">1</span></code></pre></div>
<pre><code>## [1] 665</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(backbone.splot.try3$name.short.correct))-<span class="dv">1</span></code></pre></div>
<pre><code>## [1] 86528</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(backbone.splot.try3$rank.short.correct)</code></pre></div>
<pre><code>## 
##  family   genus  higher species 
##    1880   13383    1211  105818</code></pre>
<h1 id="resolve-the-7701-additional-species-from-the-ad-hoc-version-splot2.1">Resolve the 7,701 additional species from the ad-hoc version sPlot2.1</h1>
<p>Get the species in the sPlot-July2015-version (<code>sPlot_2015_07_29_species</code>) that do not match with the backbone in the April2015_Version (sPlot 2.0, <code>sPlot_14_4_2015_species</code>) that was use in matching procedure <a href="Select%20correctly%20resolved%20names">above</a>. Apply the cleaning procedure, similar to the one used for <code>sPlot 2.0</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">miss.new &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/Analyses/Code/</span>
<span class="st">Mismatches_29_07_2015_new.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(miss.new)</code></pre></div>
<pre><code>## [1] 7701    1</code></pre>
<p>There are 7,701 names in the <code>sPlot version 29_07_2015</code> that were not in splot 2.0 (April 2015).</p>
<h2 id="name-cleaning-spelling-of-ranks-name-additions-etc.">Name cleaning (spelling of ranks, name additions etc.)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">OriginalNames &lt;-<span class="st"> </span><span class="kw">as.character</span>(miss.new$x)

OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'*'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'cf. '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'Cf. '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'['</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">']'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' x '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'Ã—'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'aff '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'('</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">')'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' cf '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">' aff. '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'c‚e'</span>, <span class="st">'ceae'</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'    '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'   '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'  '</span>, <span class="st">' '</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'x-'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'X-'</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'like '</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)  
OriginalNames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">','</span>, <span class="st">''</span>, OriginalNames, <span class="dt">fixed=</span><span class="ot">TRUE</span>)  

<span class="kw">library</span>(stringr)
firstWordWithNumbers &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">'[0-9]'</span>, <span class="kw">word</span>(OriginalNames, <span class="dv">1</span>))
numberOfWords &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">gregexpr</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">W+&quot;</span>, OriginalNames), length) +<span class="st"> </span><span class="dv">1</span>
OriginalNames[firstWordWithNumbers &amp;<span class="st"> </span>numberOfWords &gt;<span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">sapply</span>(OriginalNames[
    firstWordWithNumbers &amp;<span class="st"> </span>numberOfWords &gt;<span class="st"> </span><span class="dv">1</span>], function(x) <span class="kw">substr</span>(x, <span class="dt">start=</span><span class="kw">regexpr</span>(
                                                                         <span class="dt">pattern =</span><span class="st">' '</span>,
                                                                         <span class="dt">text=</span>x)+<span class="dv">1</span>,
                                                                  <span class="dt">stop=</span><span class="kw">nchar</span>(x)))

CleanedNames &lt;-<span class="st"> </span><span class="kw">taxname.abbr</span>(OriginalNames)
<span class="kw">write.csv</span>(CleanedNames, <span class="dt">file =</span> <span class="st">&quot;CleanedNames.csv&quot;</span>)
<span class="kw">save</span>(CleanedNames, <span class="dt">file =</span> <span class="st">&quot;CleanedNames.Rdata&quot;</span>)</code></pre></div>
<h2 id="upload-cleaned-names-to-trns">Upload cleaned names to TRNS</h2>
<p>On the <a href="http://tnrs.iplantcollaborative.org/TNRSapp.html">TNRS application webpage</a>, use all six sources (<code>TPL</code> first, <a href="Sources%20for%20name%20resolution">see procedure above</a>). Subsequently, use <code>Tropicos</code> first.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl.new &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/missing_tpl.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,
                         <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>, <span class="dt">skip =</span> <span class="dv">0</span>, <span class="dt">head =</span> T)[-<span class="dv">1</span>, ]
<span class="kw">str</span>(tnrs.tpl.new)
tnrs.tpl &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;tnrs.tpl.csv&quot;</span>)</code></pre></div>
<h2 id="select-correctly-resolved-names-tpl-first">Select correctly resolved names (TPL first)</h2>
<p>Assign index to first column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.tpl$Name_number &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="kw">length</span>(tnrs.tpl$Name_number)

tnrs.tpl &lt;-<span class="st"> </span>tnrs.tpl[,-<span class="dv">1</span>]</code></pre></div>
<h3 id="family-level-3">Family level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.family &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;family&quot;</span> &amp;
<span class="st">                               </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                               </span>tnrs.tpl$Family_score &gt;<span class="st"> </span><span class="fl">0.88</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.family)</code></pre></div>
<p>Repeat selection procedure for <code>Name_matched_rank</code> ==:</p>
<ul>
<li><code>forma</code></li>
<li><code>genus</code></li>
<li><code>infraspecies</code></li>
<li><code>species</code></li>
<li>etc.</li>
</ul>
<p>Create respective index:</p>
<h3 id="forma-level-2">Forma level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.forma &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;forma&quot;</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.forma)

### Genus level</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.genus &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                              </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> &amp;
<span class="st">                              </span>tnrs.tpl$Genus_score &gt;<span class="st"> </span><span class="fl">0.83</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus)</code></pre></div>
<h3 id="infraspecific-level">Infraspecific level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.infraspec &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;infraspecies&quot;</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.infraspec)</code></pre></div>
<h3 id="species-level-3">Species level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.species &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                </span>tnrs.tpl$Genus_score &gt;<span class="st"> </span><span class="fl">0.78</span> &amp;<span class="st"> </span>tnrs.tpl$Name_score &gt;<span class="st"> </span><span class="fl">0.93</span>),<span class="dv">1</span>]
<span class="kw">length</span>(index.species1)

index.species &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                </span>tnrs.tpl$Specific_epithet_score &gt;<span class="st"> </span><span class="fl">0.78</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species2)

index.species &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.species1, index.species2))
<span class="kw">length</span>(index.species)</code></pre></div>
<h3 id="subspecies-level-2">Subspecies level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.subspec &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.subspec)</code></pre></div>
<h3 id="variety-level-3">Variety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.variety &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span> &amp;
<span class="st">                                </span>(tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                 </span>tnrs.tpl$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)), <span class="dv">1</span>]
<span class="kw">length</span>(index.variety)</code></pre></div>
<h3 id="identifying-non-matched-species-that-are-spermatophyta-1">Identifying <code>non-matched</code> species that are spermatophyta</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.spermatophyt &lt;-<span class="st"> </span>tnrs.tpl[<span class="kw">which</span>(tnrs.tpl$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span> &amp;
<span class="st">                                     </span><span class="kw">word</span>(tnrs.tpl$Name_submitted, <span class="dv">1</span>) ==<span class="st"> &quot;Spermatophyta&quot;</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.spermatophyt)</code></pre></div>
<h3 id="identify-species-that-do-not-fulfill-the-search-criteria">Identify species that do not fulfill the search criteria</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.tpl &lt;-<span class="st"> </span><span class="kw">c</span>(index.family, index.forma, index.genus, index.species, index.subspec,
               index.variety, index.spermatophyt)
<span class="kw">length</span>((index.tpl))

tnrs.tpl.certain &lt;-<span class="st"> </span>tnrs.tpl[index.tpl,]
<span class="kw">dim</span>(tnrs.tpl.certain)
<span class="kw">save</span>(tnrs.tpl.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.certain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.tpl.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.certain.csv&quot;</span>)

tnrs.tpl.uncertain &lt;-<span class="st"> </span>tnrs.tpl[tnrs.tpl$Name_number %in%<span class="st"> </span>index.tpl ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(tnrs.tpl.uncertain)
<span class="kw">save</span>(tnrs.tpl.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.tpl.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.csv&quot;</span>)

<span class="kw">write.csv</span>(tnrs.tpl.uncertain[,<span class="dv">2</span>], <span class="dt">file =</span> <span class="st">&quot;tnrs.tpl.uncertain.upload.csv&quot;</span>)</code></pre></div>
<h2 id="select-correctly-resolved-names-tropicos-first">Select correctly resolved names (<code>Tropicos</code> first)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Downloads/tnrs.trop.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">str</span>(tnrs.trop)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(tnrs.trop)
tnrs.trop$Name_number &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="kw">length</span>(tnrs.trop$Name_number)</code></pre></div>
<h3 id="reduce-to-the-uncertain-species-from-tnrs.tpl">Reduce to the uncertain species from tnrs.tpl</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop.small &lt;-<span class="st"> </span>tnrs.trop[tnrs.trop$Name_number %in%<span class="st"> </span>index.tpl ==<span class="st"> </span>F, ]
<span class="kw">str</span>(tnrs.trop.small)

<span class="kw">save</span>(tnrs.trop.small, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.csv&quot;</span>)</code></pre></div>
<h3 id="family-level-4">Family level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.family &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;family&quot;</span> &amp;
<span class="st">                                      </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span>|
<span class="st">                                       </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span>))
                              , <span class="dv">1</span>]
<span class="kw">length</span>(index.family)</code></pre></div>
<h3 id="forma-level-3">Forma level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.forma &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;forma&quot;</span> &amp;
<span class="st">                                     </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                      </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>)),
                               <span class="dv">1</span>]
<span class="kw">length</span>(index.forma)</code></pre></div>
<h3 id="genus-level-3">Genus level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.genus &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;genus&quot;</span> &amp;
<span class="st">                                     </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                      </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span> |
<span class="st">                                      </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span>) &amp;
<span class="st">                                     </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.83</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.genus)</code></pre></div>
<h3 id="species-level-4">Species level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.species &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                       </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                        </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span> |
<span class="st">                                        </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span>) &amp;
<span class="st">                                       </span>tnrs.trop.small$Specific_epithet_score &gt;<span class="st"> </span><span class="fl">0.77</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species)

index.species2 &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;species&quot;</span> &amp;
<span class="st">                                        </span>(tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">                                         </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;Synonym&quot;</span>) &amp;
<span class="st">                                        </span>tnrs.trop.small$Genus_score &gt;<span class="st"> </span><span class="fl">0.88</span> &amp;
<span class="st">                                        </span>tnrs.trop.small$Name_score &gt;<span class="st"> </span><span class="fl">0.49</span>), <span class="dv">1</span>]
<span class="kw">length</span>(index.species2)

index.species &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(index.species1, index.species2))
<span class="kw">length</span>(index.species)</code></pre></div>
<h3 id="subspecies-level-3">Subspecies level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.subspec &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span>),
                                 <span class="dv">1</span>]
<span class="kw">length</span>(index.subspec)</code></pre></div>
<h3 id="subvariety-level">Subvariety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.subvariety &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;subvariety&quot;</span>),
                                    <span class="dv">1</span>]
<span class="kw">length</span>(index.subvariety)</code></pre></div>
<h3 id="variety-level-4">Variety level</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index.variety &lt;-<span class="st"> </span>tnrs.trop.small[<span class="kw">which</span>(tnrs.trop.small$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span> &amp;
<span class="st">                                       </span>tnrs.trop.small$Taxonomic_status ==<span class="st"> &quot;No opinion&quot;</span>),
                                 <span class="dv">1</span>]
<span class="kw">length</span>(index.variety)

index.all &lt;-<span class="st"> </span><span class="kw">c</span>(index.family, index.genus, index.species, index.subspec, index.variety,
               index.subvariety)
<span class="kw">length</span>((index.all))

tnrs.trop.small.certain &lt;-<span class="st"> </span>tnrs.trop.small[tnrs.trop.small$Name_number %in%
<span class="st">                                           </span>index.all ==<span class="st"> </span>T,]
<span class="kw">dim</span>(tnrs.trop.small.certain)
<span class="kw">save</span>(tnrs.trop.small.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.certain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small.certain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.certain.csv&quot;</span>)

tnrs.trop.small.uncertain &lt;-<span class="st"> </span>tnrs.trop.small[tnrs.trop.small$Name_number %in%
<span class="st">                                             </span>index.all ==<span class="st"> </span>F, ]
<span class="kw">dim</span>(tnrs.trop.small.uncertain)
<span class="kw">save</span>(tnrs.trop.small.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.uncertain.Rdata&quot;</span>)
<span class="kw">write.csv</span>(tnrs.trop.small.uncertain, <span class="dt">file =</span> <span class="st">&quot;tnrs.trop.small.uncertain.csv&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain &lt;-<span class="st"> </span><span class="kw">rbind</span>(tnrs.tpl.certain, tnrs.trop.small.certain)
<span class="kw">str</span>(backbone.tpl.trop.certain)</code></pre></div>
<h3 id="kick-out-some-columns-that-should-not-be-in-the-final-backbone">Kick out some columns that should not be in the final backbone</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.2</span> &lt;-<span class="st"> </span>backbone.tpl.trop.certain[ ,<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">12</span>:<span class="dv">17</span>,<span class="dv">25</span>:<span class="dv">35</span>)]
<span class="kw">colnames</span>(backbone.tpl.trop.certain<span class="fl">.2</span>)</code></pre></div>
<h3 id="add-extra-columns-to-match-the-old-backbone">Add extra columns to match the old backbone</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.2</span>$Manual.matching &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$Status.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$name.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$rank.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$family.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$name.short.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$rank.short.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$names.sPlot.TRY &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$names.corr.string &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.tpl.trop.certain<span class="fl">.2</span>$sPlot.TRY &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h3 id="bring-names-in-the-same-order-as-in-the-old-backbone">Bring names in the same order as in the old backbone</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span> &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">31</span>:<span class="dv">33</span>,<span class="dv">2</span>:<span class="dv">30</span>)]
<span class="kw">match</span>(<span class="kw">colnames</span>(backbone.tpl.trop.certain<span class="fl">.3</span>), <span class="kw">colnames</span>(backbone.splot.try3))
<span class="kw">identical</span>(<span class="kw">colnames</span>(backbone.tpl.trop.certain<span class="fl">.3</span>), <span class="kw">colnames</span>(backbone.splot.try3)) <span class="co"># TRUE</span></code></pre></div>
<h2 id="fill-in-empty-columns">Fill in empty columns</h2>
<h3 id="status.correct"><code>Status.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$Status.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$Status.correct[index] &lt;-<span class="st"> &quot;No suitable matches found.&quot;</span></code></pre></div>
<p>Rename <code>no.opinion</code> to <code>unresolved</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Status.correct ==<span class="st"> &quot;No opinion&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$Status.correct[index] &lt;-<span class="st"> &quot;Unresolved&quot;</span></code></pre></div>
<h3 id="name.correct"><code>Name.correct</code></h3>
<p>Use &quot;Accepted name&quot;, if unresolved take &quot;Name_matched&quot;)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$name.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Accepted_name
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Synonym&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$name.correct[index] &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$
<span class="st">    </span>Name_matched[index]
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$name.correct[index] &lt;-<span class="st"> &quot;No suitable matches found.&quot;</span></code></pre></div>
<h3 id="rank.correct"><code>Rank.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Accepted_name_rank
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Synonym&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct[index] &lt;-
<span class="st">    </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched_rank[index]
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct[index] &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h3 id="family.correct"><code>Family.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$family.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$
<span class="st">    </span>Accepted_name_family
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Accepted&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Taxonomic_status !=<span class="st"> &quot;Synonym&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$family.correct[index] &lt;-
<span class="st">    </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched_accepted_family[index]
index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched ==<span class="st"> &quot;No suitable matches found.&quot;</span>)
backbone.tpl.trop.certain<span class="fl">.3</span>$family.correct[index] &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h3 id="name.short.correct"><code>Name.short.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$name.short.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$name.correct

index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;subspecies&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;variety&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;forma&quot;</span>|
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;subvariety&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched_rank ==<span class="st"> &quot;subspecies&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$Name_matched_rank ==<span class="st"> &quot;variety&quot;</span>)

<span class="kw">library</span>(stringr)
backbone.tpl.trop.certain<span class="fl">.3</span>$name.short.correct[index] &lt;-
<span class="st">    </span><span class="kw">word</span>(<span class="dt">string =</span> backbone.tpl.trop.certain<span class="fl">.3</span>$name.short.correct[index], <span class="dt">start =</span> <span class="dv">1</span>, <span class="dt">end =</span> <span class="dv">2</span>)</code></pre></div>
<h3 id="rank.short.correct"><code>Rank.short.correct</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop.certain<span class="fl">.3</span>$rank.short.correct &lt;-<span class="st"> </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct

index &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;subspecies&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;variety&quot;</span> |
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;forma&quot;</span>|
<span class="st">               </span>backbone.tpl.trop.certain<span class="fl">.3</span>$rank.correct ==<span class="st"> &quot;subvariety&quot;</span>)

backbone.tpl.trop.certain<span class="fl">.3</span>$rank.short.correct[index] &lt;-<span class="st"> &quot;species&quot;</span></code></pre></div>
<h3 id="kick-out-some-columns-that-should-not-be-in-the-final-backbone-1">Kick out some columns that should not be in the final backbone</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop.small.uncertain<span class="fl">.2</span> &lt;-<span class="st"> </span>tnrs.trop.small.uncertain[ ,<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">12</span>:<span class="dv">17</span>,<span class="dv">25</span>:<span class="dv">35</span>)]
<span class="kw">colnames</span>(tnrs.trop.small.uncertain<span class="fl">.2</span>)</code></pre></div>
<p>Add extra columns to match the old backbone</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop.small.uncertain<span class="fl">.2</span>$Manual.matching &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$Status.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$name.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$rank.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$family.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$name.short.correct &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$rank.short.correct &lt;-<span class="st"> </span><span class="ot">NA</span>

tnrs.trop.small.uncertain<span class="fl">.2</span>$names.sPlot.TRY &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$names.corr.string &lt;-<span class="st"> </span><span class="ot">NA</span>
tnrs.trop.small.uncertain<span class="fl">.2</span>$sPlot.TRY &lt;-<span class="st"> </span><span class="ot">NA</span></code></pre></div>
<h4 id="bring-names-in-the-same-order-as-in-the-old-backbone-1">Bring names in the same order as in the old backbone</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tnrs.trop.small.uncertain<span class="fl">.3</span> &lt;-<span class="st"> </span>tnrs.trop.small.uncertain<span class="fl">.2</span>[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">31</span>:<span class="dv">33</span>,<span class="dv">2</span>:<span class="dv">30</span>)]
<span class="kw">match</span>(<span class="kw">colnames</span>(tnrs.trop.small.uncertain<span class="fl">.3</span>), <span class="kw">colnames</span>(backbone.splot.try3))
<span class="kw">identical</span>(<span class="kw">colnames</span>(tnrs.trop.small.uncertain<span class="fl">.3</span>), <span class="kw">colnames</span>(backbone.splot.try3)) <span class="co"># TRUE</span>

tnrs.trop.small.uncertain<span class="fl">.3</span>[,<span class="kw">c</span>(<span class="dv">28</span>:<span class="dv">29</span>)] &lt;-<span class="st"> &quot;No suitable matches found.&quot;</span></code></pre></div>
<h4 id="merge-with-backbone-for-certain-species">Merge with backbone for certain species</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop &lt;-<span class="st"> </span><span class="kw">rbind</span>(backbone.tpl.trop.certain<span class="fl">.3</span>, tnrs.trop.small.uncertain<span class="fl">.3</span>)
<span class="kw">head</span>(backbone.tpl.trop)
<span class="kw">dim</span>(backbone.tpl.trop)
<span class="kw">write.csv</span>(backbone.tpl.trop, <span class="dt">file =</span> <span class="st">&quot;backbone.tpl.trop.csv&quot;</span>)
backbone.tpl.trop &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;backbone.tpl.trop.csv&quot;</span>)</code></pre></div>
<h3 id="fill-in-remaining-first-three-columns">Fill in remaining first three columns</h3>
<p>Assing all entries to sPlot (&quot;S&quot;)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop$sPlot.TRY &lt;-<span class="st"> &quot;S&quot;</span></code></pre></div>
<p>Assing all entries to sPlot-version 2.1 (&quot;sPlot2b&quot;)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop$Manual.matching &lt;-<span class="st"> &quot;sPlot2b&quot;</span>
miss.clean<span class="fl">.2</span> &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dv">1</span>:<span class="kw">length</span>(miss.clean[,<span class="dv">1</span>]), <span class="kw">as.data.frame</span>(miss.clean))
<span class="kw">colnames</span>(miss.clean<span class="fl">.2</span>)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;Name_number&quot;</span></code></pre></div>
<h3 id="join-the-trns-results">Join the TRNS results</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.tpl.trop &lt;-<span class="st"> </span><span class="kw">left_join</span>(miss.clean<span class="fl">.2</span>, backbone.tpl.trop, <span class="dt">by =</span> <span class="st">&quot;Name_number&quot;</span>)
<span class="kw">head</span>(backbone.tpl.trop)

<span class="kw">colnames</span>(backbone.tpl.trop)

backbone.tpl.trop<span class="fl">.2</span> &lt;-<span class="st"> </span>backbone.tpl.trop[,-<span class="kw">c</span>(<span class="dv">4</span>:<span class="dv">6</span>)]
<span class="kw">colnames</span>(backbone.tpl.trop<span class="fl">.2</span>)[<span class="dv">2</span>:<span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;names.sPlot.TRY&quot;</span>, <span class="st">&quot;names.corr.string&quot;</span>)
<span class="kw">match</span>(<span class="kw">colnames</span>(backbone.tpl.trop<span class="fl">.2</span>), <span class="kw">colnames</span>(backbone.splot.try3))
<span class="kw">identical</span>(<span class="kw">colnames</span>(backbone.tpl.trop<span class="fl">.2</span>), <span class="kw">colnames</span>(backbone.splot.try3))

<span class="kw">write.csv</span>(backbone.tpl.trop<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;backbone.tpl.trop.2.csv&quot;</span>)
<span class="kw">save</span>(backbone.tpl.trop<span class="fl">.2</span>, <span class="dt">file =</span> <span class="st">&quot;backbone.tpl.trop.2.Rdata&quot;</span>)</code></pre></div>
<h3 id="merge-additional-species-with-the-big-backbone-for-splot-2.0-and-try-3.0">Merge additional species with the big backbone for <code>sPlot 2.0</code> and <code>TRY 3.0</code></h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.splot2b.try3 &lt;-<span class="st"> </span><span class="kw">rbind</span>(backbone.splot.try3, backbone.tpl.trop<span class="fl">.2</span>)
<span class="kw">dim</span>(backbone.splot2b.try3)</code></pre></div>
<p>Check whether the new backbone matches the <code>sPlot 2.1</code> species data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(splot.species)
<span class="kw">length</span>(<span class="kw">unique</span>(splot.species$Matched_concept))
<span class="kw">which</span>(<span class="kw">unique</span>(splot.species$Matched_concept) %in%<span class="st"> </span>backbone.splot2b.try3$names.sPlot.TRY ==<span class="st"> </span>F)
backbone.splot2<span class="fl">.1</span>.try3 &lt;-<span class="st"> </span>backbone.splot2b.try3
<span class="kw">write.csv</span>(backbone.splot2<span class="fl">.1</span>.try3, <span class="dt">file =</span> <span class="st">&quot;backbone.splot2.1.try3.csv&quot;</span>)
<span class="kw">save</span>(backbone.splot2<span class="fl">.1</span>.try3, <span class="dt">file =</span> <span class="st">&quot;backbone.splot2.1.try3.Rdata&quot;</span>)</code></pre></div>
<h1 id="tag-vascular-species-in-backbone.splot2.1.try3">Tag vascular species in <code>backbone.splot2.1.try3</code></h1>
<p>Because sPlot, and to some extent TRY, contain a bunch of names belonging to non-vascular, those need to be tagged.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;../backbone.splot2.1.try3.is.vascular.Rdata&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(backbone.splot2<span class="fl">.1</span>.try3)</code></pre></div>
<pre><code>##  [1] &quot;Name_number&quot;                  &quot;names.sPlot.TRY&quot;             
##  [3] &quot;names.corr.string&quot;            &quot;sPlot.TRY&quot;                   
##  [5] &quot;Name_submitted&quot;               &quot;Overall_score&quot;               
##  [7] &quot;Name_matched&quot;                 &quot;Name_matched_rank&quot;           
##  [9] &quot;Name_score&quot;                   &quot;Family_score&quot;                
## [11] &quot;Name_matched_accepted_family&quot; &quot;Genus_matched&quot;               
## [13] &quot;Genus_score&quot;                  &quot;Specific_epithet_matched&quot;    
## [15] &quot;Specific_epithet_score&quot;       &quot;Unmatched_terms&quot;             
## [17] &quot;Taxonomic_status&quot;             &quot;Accepted_name&quot;               
## [19] &quot;Accepted_name_author&quot;         &quot;Accepted_name_rank&quot;          
## [21] &quot;Accepted_name_url&quot;            &quot;Accepted_name_species&quot;       
## [23] &quot;Accepted_name_family&quot;         &quot;Selected&quot;                    
## [25] &quot;Source&quot;                       &quot;Warnings&quot;                    
## [27] &quot;Manual.matching&quot;              &quot;Status.correct&quot;              
## [29] &quot;name.correct&quot;                 &quot;rank.correct&quot;                
## [31] &quot;family.correct&quot;               &quot;name.short.correct&quot;          
## [33] &quot;rank.short.correct&quot;           &quot;is.vascular.species&quot;         
## [35] &quot;sPlot2.1.TRY&quot;</code></pre>
<h2 id="get-families">Get families</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fam &lt;-<span class="st"> </span><span class="kw">unique</span>(backbone.splot2<span class="fl">.1</span>.try3$family.correct)
<span class="kw">head</span>(fam)</code></pre></div>
<pre><code>## [1] &quot;&quot;             &quot;Asparagaceae&quot; &quot;Poaceae&quot;      &quot;Fabaceae&quot;    
## [5] &quot;Polygalaceae&quot; &quot;Orchidaceae&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gbiffam &lt;-<span class="st"> </span><span class="kw">sapply</span>(fam[<span class="dv">2</span>:<span class="dv">5</span>], function(x) <span class="kw">name_usage</span>(<span class="dt">name=</span>x,
                                                   <span class="dt">rank =</span> <span class="st">'FAMILY,'</span>, <span class="dt">limit =</span> <span class="dv">1</span>)$data$phylum)
gbiffam[<span class="kw">which</span>(<span class="kw">sapply</span>(gbiffam, function(x) <span class="kw">is.null</span>(x)))] &lt;-<span class="st"> 'unknown'</span>
fam$phylum &lt;-<span class="st"> </span><span class="kw">unlist</span>(gbiffam, <span class="dt">use.names =</span> <span class="ot">TRUE</span>)
<span class="kw">table</span>(fam$phylum)
<span class="kw">write.csv</span>(fam, <span class="dt">file=</span><span class="st">'family_affiliation_gbif.csv'</span>)
family_affiliation_gbif &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">'../Florian_TaxStand/family_affiliation_gbif.csv'</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(family_affiliation_gbif$phylum)</code></pre></div>
<pre><code>## 
## Anthocerotophyta       Ascomycota    Basidiomycota        Bryophyta 
##                1               35               15               57 
##       Charophyta      Chlorophyta    Cyanobacteria      Glaucophyta 
##                3                6                2                1 
##  Marchantiophyta       Ochrophyta       Rhodophyta     Tracheophyta 
##               45               17                8              474 
##          unknown 
##                5</code></pre>
<h3 id="add-column-is.vascular.species-and-set-all-families-that-correspond-to-tracheophyta-vascular-plants-true">Add column <code>&quot;is.vascular.species&quot;</code> and set all families that correspond to <code>Tracheophyta</code> (<code>vascular plants == TRUE</code>)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">is.na</span>(backbone.splot2<span class="fl">.1</span>.try3$is.vascular.species))</code></pre></div>
<pre><code>## 
##  FALSE   TRUE 
## 123589   7013</code></pre>
<p>Select families that belong to <code>Tracheophyta</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fam.trach &lt;-<span class="st"> </span>family_affiliation_gbif$Var1[family_affiliation_gbif$phylum ==<span class="st"> &quot;Tracheophyta&quot;</span>]
<span class="kw">dim</span>(fam.trach)
ind.vasc &lt;-<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3$family.correct %in%<span class="st"> </span>fam.trach
backbone.splot2<span class="fl">.1</span>.try3$is.vascular.species &lt;-<span class="st"> </span>ind.vasc
backbone.splot2<span class="fl">.1</span>.try3$is.vascular.species[backbone.splot2<span class="fl">.1</span>.try3$is.vascular.species ==
<span class="st">                                           </span><span class="ot">FALSE</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>
backbone.splot2<span class="fl">.1</span>.try3.is.vascular &lt;-<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3</code></pre></div>
<p>To obtain proper stats for <code>sPlot2.1</code>, kick out (or tag) the species that were only found in <code>sPlot 2.0</code> but not in <code>sPlot 2.1</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splot.species.july2015 &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/</span>
<span class="st">Analyses/Data/Species/sPlot/sPlot_2015_07_29/sPlot_2015_07_29_species.csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">gc</span>()

splot.species.july2015.spec &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">unique</span>(splot.species.july2015$Matched.concept))
<span class="kw">str</span>(splot.species.july2015.spec)
<span class="kw">write.csv</span>(splot.species.july2015.spec, <span class="dt">file =</span> <span class="st">&quot;splot.species.july2015.spec.csv&quot;</span>)</code></pre></div>
<p>There are 86,432 unique, uncorrected names in <code>sPlot 2.1</code>. Check whether all of these names are in the new backbone?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;backbone.splot2.1.try3.is.vascular.Rdata&quot;</span>)</code></pre></div>
<p>Export small backbone:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.splot2<span class="fl">.1</span>.try3.small &lt;-<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">32</span>)]
<span class="kw">write.csv</span>(backbone.splot2<span class="fl">.1</span>.try3.small, <span class="dt">file =</span> <span class="st">&quot;backbone.splot2.1.try3.small.csv&quot;</span>)

<span class="kw">load</span>(<span class="st">&quot;backbone.splot2.1.try3.Rdata&quot;</span>)
ind &lt;-<span class="st"> </span>(splot.species.july2015.spec %in%<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3$names.sPlot.TRY)
splot.species.july2015.spec[ind==F]
<span class="kw">table</span>(ind)</code></pre></div>
<p>But there are only 86,427 names here. Five species are missing! Why?</p>
<p>The following species are in <code>sPlot2.1</code> (<code>splot.species.july2015.spec</code>) but not in the backbone (<code>backbone.splot2.1.try</code>):</p>
<ul>
<li>&quot;Strauch like Oraniquelocarpus&quot;</li>
<li>&quot;Buchenavia [GUNDINGA]&quot;</li>
<li>&quot;[ms554 Triumfetta Ahorn]&quot;</li>
<li>&quot;Dolichos holosericea&quot;</li>
<li>&quot;[Aeschynomene latifolia]&quot;</li>
</ul>
<p>Actually, they are in the backbone but just spelled differently, e.g. <code>Strauch like &quot;&quot;Oraniquelocarpus&quot;&quot;</code> or <code>[ms554 Triumfetta &quot;Ahorn&quot;]</code>. Probably those names got messed up, generate a vector of that 5 species that is consistent with the spelling used in the <code>missing-species-list</code> and in the final backbone.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">miss7701 &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;/home/oliver/Dokumente/PhD/PostPhD/IDiv/sDiv/sPlot/Analyses/Code/</span>
<span class="st">Mismatches_29_07_2015_new.csv&quot;</span>, <span class="dt">stringsAsFactors =</span> F)
vec5 &lt;-<span class="st"> </span>miss7701[<span class="kw">c</span>(<span class="dv">49</span>, <span class="dv">3165</span>, <span class="dv">5793</span>, <span class="dv">5814</span>, <span class="dv">5823</span>), <span class="dv">1</span>]</code></pre></div>
<p>Add these five species to the <code>sPlot 2.1</code> species list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">splot.species.july2015.spec<span class="fl">.5</span> &lt;-<span class="st"> </span><span class="kw">c</span>(splot.species.july2015.spec, vec5)</code></pre></div>
<p>In the big backbone, identify species that are in &quot;S&quot; and &quot;ST&quot; and that are in our <code>sPlot 2.1</code> species list (should be 86,432).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind &lt;-<span class="st"> </span>(splot.species.july2015.spec<span class="fl">.5</span> %in%<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3$names.sPlot.TRY)
splot.species.july2015 &lt;-<span class="st"> </span>splot.species.july2015.spec<span class="fl">.5</span>[ind]
ind &lt;-<span class="st"> </span>(splot.species.july2015 %in%<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3$names.sPlot.TRY)
<span class="kw">table</span>(ind)</code></pre></div>
<p>Great, these are the species in <code>sPlot 2.1</code>.</p>
<h3 id="select-species-in-backbone-that-are-in-splot2.1-splot.species.july2015">Select species in backbone that are in <code>sPlot2.1</code> (<code>splot.species.july2015</code>)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind.splot2<span class="fl">.1</span> &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.splot2<span class="fl">.1</span>.try3$names.sPlot.TRY %in%<span class="st"> </span>splot.species.july2015)</code></pre></div>
<p>There are 86,432 species in TRY, that are both only in TRY as well as in sPlot and TRY.</p>
<h3 id="select-species-that-are-in-try">Select species that are in TRY</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind.try &lt;-<span class="st"> </span><span class="kw">which</span>(backbone.splot2<span class="fl">.1</span>.try3$sPlot.TRY==<span class="st">&quot;T&quot;</span> |<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3$sPlot.TRY==<span class="st">&quot;ST&quot;</span>)</code></pre></div>
<p>60,273 species that are in TRY.</p>
<h3 id="species-that-are-both-in-splot-and-try">Species that are both in sPlot and TRY</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">intersect_all &lt;-<span class="st"> </span>function(a,b,...){
  <span class="kw">Reduce</span>(intersect, <span class="kw">list</span>(a,b,...))
}

inter &lt;-<span class="st"> </span><span class="kw">intersect_all</span>(ind.splot2<span class="fl">.1</span>, ind.try)</code></pre></div>
<p>24,844 names shared between sPlot and TRY.</p>
<p>Check whether the three numbers above are correct, and add new column <code>sPlot2.1.TRY</code> to backbone, that tags the entries that belong to <code>sPlot 2.1</code> and <code>TRY 3.0</code>, respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">backbone.splot2<span class="fl">.1</span>.try3$ST &lt;-<span class="st"> </span><span class="ot">NA</span>

backbone.splot2<span class="fl">.1</span>.try3$ST[ind.splot2<span class="fl">.1</span>] &lt;-<span class="st"> &quot;S&quot;</span>
backbone.splot2<span class="fl">.1</span>.try3$ST[ind.try] &lt;-<span class="st"> &quot;T&quot;</span>
backbone.splot2<span class="fl">.1</span>.try3$ST[inter] &lt;-<span class="st"> &quot;ST&quot;</span>
<span class="kw">colnames</span>(backbone.splot2<span class="fl">.1</span>.try3)[<span class="dv">35</span>] &lt;-<span class="st"> &quot;sPlot2.1.TRY&quot;</span>
<span class="kw">save</span>(backbone.splot2<span class="fl">.1</span>.try3, <span class="dt">file =</span> <span class="st">&quot;backbone.splot2.1.try3.is.vascular.Rdata&quot;</span>)</code></pre></div>
<h1 id="statistics-for-backbone-combining-names-in-splot2.1-and-try3.0">Statistics for backbone combining names in <code>sPlot2.1</code> and <code>TRY3.0</code></h1>
<h2 id="all-taxon-name-entries">All taxon name entries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;../backbone.splot2.1.try3.is.vascular.Rdata&quot;</span>)</code></pre></div>
<p><strong>How many entries in the backbone were only found in the old sPlot-version 2.0 (<code>sPlot_14_4_2015</code>) but not in <code>sPlot 2.1</code> <code>sPlot_2015_07_19</code>?</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(!<span class="kw">is.na</span>(backbone.splot2<span class="fl">.1</span>.try3$sPlot2<span class="fl">.1</span>.TRY))</code></pre></div>
<pre><code>## 
##  FALSE   TRUE 
##   8741 121861</code></pre>
<p>There are 8,741 entries that do not belong to <code>sPlot 2.1</code>. In total 121,861 (unstandardized) name entries in <code>sPlot 2.1</code> and <code>TRY 3.0</code> combined. Index those species that were exclusiely found in the sPlot 2.0 but not in version 2.1, and remove them from the backbone (<code>back2.1</code>) to update the match statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind2<span class="fl">.1</span> &lt;-<span class="st"> </span>!<span class="kw">is.na</span>(backbone.splot2<span class="fl">.1</span>.try3$sPlot2<span class="fl">.1</span>.TRY)
back2<span class="fl">.1</span> &lt;-<span class="st"> </span>backbone.splot2<span class="fl">.1</span>.try3[ind2<span class="fl">.1</span>, ]</code></pre></div>
<p><strong>Database affiliations (<code>sPlot 2.1</code> and <code>TRY 3.0</code>).</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(back2<span class="fl">.1</span>$sPlot2<span class="fl">.1</span>.TRY)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) name entries</span>
<span class="st">unique to, or shared between TRY (S) and sPlot (T).&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">S</th>
<th align="right">ST</th>
<th align="right">T</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">61588</td>
<td align="right">24844</td>
<td align="right">35429</td>
</tr>
</tbody>
</table>
<p>60,273 of the total number of entries belong to TRY (incl. names that are only in TRY as well as in sPlot and TRY). 86,432 name entries belong to sPlot (incl. names that are only in TRY as well as in sPlot and TRY).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(back2<span class="fl">.1</span>$Manual.matching)</code></pre></div>
<pre><code>## 
## sPlot2b       x 
##    7694    1432</code></pre>
<p>1,432 name entries were matched manually (see above).</p>
<p><strong>Taxonomic ranks:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(back2<span class="fl">.1</span>$rank.short.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) name entries</span>
<span class="st">per taxonomic rank.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">family</th>
<th align="right">genus</th>
<th align="right">higher</th>
<th align="right">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1745</td>
<td align="right">12373</td>
<td align="right">1020</td>
<td align="right">105777</td>
</tr>
</tbody>
</table>
<p><strong>Taxonomic status:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(back2<span class="fl">.1</span>$Status.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) name entries</span>
<span class="st">that correspond to `Accepted`, `Synonyms` or Unresolved species, respecively.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">Accepted</th>
<th align="right">No suitable matches found.</th>
<th align="right">Synonym</th>
<th align="right">Unresolved</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">94068</td>
<td align="right">1902</td>
<td align="right">21028</td>
<td align="right">4863</td>
</tr>
</tbody>
</table>
<p><strong>Total number of unique standardized taxon names and families:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>$name.short.correct))-<span class="dv">1</span> <span class="co"># minus 1 for NA</span></code></pre></div>
<pre><code>## [1] 86760</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>$family.correct))-<span class="dv">2</span> <span class="co"># minus 2 for &quot;&quot; and NA</span></code></pre></div>
<pre><code>## [1] 663</code></pre>
<p><strong>Number of entries corresponding to vascular plant species:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(back2<span class="fl">.1</span>$is.vascular.species)</code></pre></div>
<pre><code>## 
##   TRUE 
## 115678</code></pre>
<p><strong>Number of duplicated entries after taxonomic standardization:</strong> Frequency of original (non-standardized) species names per resolved (non-standardized) name (excluding non-vascular and non-matched species).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.count &lt;-<span class="st"> </span>back2<span class="fl">.1</span> %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct)) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">group_by</span>(name.short.correct) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">summarise</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">arrange</span>(<span class="kw">desc</span>(n))

<span class="kw">kable</span>(df.count[<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">3</span>, <span class="dv">21</span>:<span class="dv">25</span>, <span class="dv">101</span>:<span class="dv">110</span>), ], , <span class="dt">caption =</span> <span class="st">&quot;Number of unresolved, original name</span>
<span class="st">enties per resolved name.&quot;</span>)</code></pre></div>
<h2 id="based-on-unique-standardized-names">Based on <code>unique</code> standardized names</h2>
<p>Generate version of the backbone that only includes the unique resolved names in <code>name.short.correct</code>, and for the non-unique names, the first rows of duplicated name:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[!<span class="kw">duplicated</span>(back2<span class="fl">.1</span>$name.short.correct), ]</code></pre></div>
<p>Remove the first entry, which is NA:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni[-<span class="dv">1</span>, ]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>.uni$name.short.correct))</code></pre></div>
<pre><code>## [1] 86761</code></pre>
<p>There are 86,760 unique taxon names the in backbone. Exclude the non-vascular plant and non-matching taxon names:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct))</code></pre></div>
<p><code>df.uni</code> had two names less than <code>df.count</code>, as they were accidentially tagged as non-vascular species names. Resolve that issue:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">miss &lt;-<span class="st"> </span>df.count$name.short.correct[(df.count$name.short.correct %in%<span class="st"> </span>df.uni$name.short.correct) ==<span class="st"> </span>F]
<span class="kw">which</span>(df.count$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))
<span class="kw">which</span>(df.uni$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))
df.count[<span class="kw">c</span>(<span class="dv">20319</span>, <span class="dv">31478</span>), ]
ind &lt;-<span class="st"> </span><span class="kw">which</span>(back2<span class="fl">.1</span>$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>[ind, <span class="dv">31</span>:<span class="dv">34</span>]</code></pre></div>
<pre><code>##       family.correct name.short.correct rank.short.correct
## 8853    Brassicaceae  Arabis stellulata            species
## 8932    Brassicaceae  Arabis stellulata            species
## 30191  Ranunculaceae Coptidium pallasii            species
## 96184  Ranunculaceae Coptidium pallasii            species
##       is.vascular.species
## 8853                 TRUE
## 8932                 TRUE
## 30191                TRUE
## 96184                TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>[ind[<span class="dv">1</span>], <span class="dv">31</span>:<span class="dv">34</span>] &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[ind[<span class="dv">2</span>], <span class="dv">31</span>:<span class="dv">34</span>]
back2<span class="fl">.1</span>[ind[<span class="dv">3</span>], <span class="dv">31</span>:<span class="dv">34</span>] &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[ind[<span class="dv">4</span>], <span class="dv">31</span>:<span class="dv">34</span>]

df.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct))</code></pre></div>
<p><strong>Now, run the stats for unique resolved names (excluding non-vascular and non-matching taxa):</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(df.uni$name.short.correct)</code></pre></div>
<pre><code>## [1] 83677</code></pre>
<p>There are 83,679 unique (non-vascular plant) taxon names:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$sPlot2<span class="fl">.1</span>.TRY)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular plant</span>
<span class="st">taxon names per unique to, and shared between TRY (S) and sPlot (T).&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">S</th>
<th align="right">ST</th>
<th align="right">T</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">34105</td>
<td align="right">20414</td>
<td align="right">29158</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(df.uni$Manual.matching)</code></pre></div>
<pre><code>## 
## sPlot2b       x 
##    1637     490</code></pre>
<p>490 of those names were matched manually (see above).</p>
<p><strong>Taxonomic ranks:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$rank.short.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) name entries</span>
<span class="st">per taxonomic rank.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">family</th>
<th align="right">genus</th>
<th align="right">higher</th>
<th align="right">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">152</td>
<td align="right">4343</td>
<td align="right">13</td>
<td align="right">79169</td>
</tr>
</tbody>
</table>
<p><strong>Taxonomic status:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$Status.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) name entries that</span>
<span class="st">correspond to `Accepted`, `Synonyms` or Unresolved species, respecively.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">Accepted</th>
<th align="right">Synonym</th>
<th align="right">Unresolved</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">69173</td>
<td align="right">9998</td>
<td align="right">4506</td>
</tr>
</tbody>
</table>
<p><strong>Total number of unique standardized taxon names and families:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>$name.short.correct))-<span class="dv">1</span> <span class="co"># minus 1 for NA</span></code></pre></div>
<pre><code>## [1] 86760</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>$family.correct))-<span class="dv">2</span> <span class="co"># minus 2 for &quot;&quot; and NA</span></code></pre></div>
<pre><code>## [1] 663</code></pre>
<p><strong>Number of entries corresponding to vascular plant species:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(back2<span class="fl">.1</span>$is.vascular.species)</code></pre></div>
<pre><code>## 
##   TRUE 
## 115678</code></pre>
<p><strong>Number of duplicated entries after taxonomic standardization:</strong> Frequency of original (non-standardized) species names per resolved (non-standardized) name (excluding non-vascular and non-matched species).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.count &lt;-<span class="st"> </span>back2<span class="fl">.1</span> %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct)) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">group_by</span>(name.short.correct) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">summarise</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%
<span class="st">    </span>dplyr::<span class="kw">arrange</span>(<span class="kw">desc</span>(n))

<span class="kw">kable</span>(df.count[<span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">3</span>, <span class="dv">21</span>:<span class="dv">25</span>, <span class="dv">101</span>:<span class="dv">110</span>), ], , <span class="dt">caption =</span> <span class="st">&quot;Number of unresolved, original</span>
<span class="st">name enties per resolved name.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">name.short.correct</th>
<th align="right">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Poaceae</td>
<td align="right">254</td>
</tr>
<tr class="even">
<td align="left">Lauraceae</td>
<td align="right">175</td>
</tr>
<tr class="odd">
<td align="left">Asteraceae</td>
<td align="right">146</td>
</tr>
<tr class="even">
<td align="left">Myrcia</td>
<td align="right">47</td>
</tr>
<tr class="odd">
<td align="left">Taraxacum</td>
<td align="right">42</td>
</tr>
<tr class="even">
<td align="left">Pouteria</td>
<td align="right">41</td>
</tr>
<tr class="odd">
<td align="left">Malpighiaceae</td>
<td align="right">39</td>
</tr>
<tr class="even">
<td align="left">Guarea</td>
<td align="right">37</td>
</tr>
<tr class="odd">
<td align="left">Annonaceae</td>
<td align="right">17</td>
</tr>
<tr class="even">
<td align="left">Antimima</td>
<td align="right">17</td>
</tr>
<tr class="odd">
<td align="left">Brosimum</td>
<td align="right">17</td>
</tr>
<tr class="even">
<td align="left">Centaurium erythraea</td>
<td align="right">17</td>
</tr>
<tr class="odd">
<td align="left">Chrysophyllum</td>
<td align="right">17</td>
</tr>
<tr class="even">
<td align="left">Cinnamomum</td>
<td align="right">17</td>
</tr>
<tr class="odd">
<td align="left">Elymus hispidus</td>
<td align="right">17</td>
</tr>
<tr class="even">
<td align="left">Euphorbia</td>
<td align="right">17</td>
</tr>
<tr class="odd">
<td align="left">Lamium galeobdolon</td>
<td align="right">17</td>
</tr>
<tr class="even">
<td align="left">Mollinedia</td>
<td align="right">17</td>
</tr>
</tbody>
</table>
<h2 id="based-on-unique-standardized-names-1">Based on <code>unique</code> standardized names</h2>
<p>Generate version of the backbone that only includes the unique resolved names in <code>name.short.correct</code>, and for the non-unique names, the first rows of duplicated name:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[!<span class="kw">duplicated</span>(back2<span class="fl">.1</span>$name.short.correct), ]</code></pre></div>
<p>Remove the first entry, which is NA:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni[-<span class="dv">1</span>, ]
<span class="kw">length</span>(<span class="kw">unique</span>(back2<span class="fl">.1</span>.uni$name.short.correct))</code></pre></div>
<p>There are 86,760 unique taxon names the in backbone. Exclude the non-vascular plant and non-matching taxon names:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct))</code></pre></div>
<p><code>df.uni</code> had two names less than <code>df.count</code>, as they were accidentially tagged as non-vascular species names. Correct that:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">miss &lt;-<span class="st"> </span>df.count$name.short.correct[(df.count$name.short.correct %in%<span class="st"> </span>df.uni$name.short.correct) ==<span class="st"> </span>F]
<span class="kw">which</span>(df.count$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))
<span class="kw">which</span>(df.uni$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))
df.count[<span class="kw">c</span>(<span class="dv">20319</span>, <span class="dv">31478</span>), ]
ind &lt;-<span class="st"> </span><span class="kw">which</span>(back2<span class="fl">.1</span>$name.short.correct %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Arabis stellulata&quot;</span>, <span class="st">&quot;Coptidium pallasii&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>[ind, <span class="dv">31</span>:<span class="dv">34</span>]</code></pre></div>
<pre><code>##       family.correct name.short.correct rank.short.correct
## 8853    Brassicaceae  Arabis stellulata            species
## 8932    Brassicaceae  Arabis stellulata            species
## 30191  Ranunculaceae Coptidium pallasii            species
## 96184  Ranunculaceae Coptidium pallasii            species
##       is.vascular.species
## 8853                 TRUE
## 8932                 TRUE
## 30191                TRUE
## 96184                TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">back2<span class="fl">.1</span>[ind[<span class="dv">1</span>], <span class="dv">31</span>:<span class="dv">34</span>] &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[ind[<span class="dv">2</span>], <span class="dv">31</span>:<span class="dv">34</span>]
back2<span class="fl">.1</span>[ind[<span class="dv">3</span>], <span class="dv">31</span>:<span class="dv">34</span>] &lt;-<span class="st"> </span>back2<span class="fl">.1</span>[ind[<span class="dv">4</span>], <span class="dv">31</span>:<span class="dv">34</span>]

df.uni &lt;-<span class="st"> </span>back2<span class="fl">.1</span>.uni %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct))</code></pre></div>
<p><strong>Now, run the stats for unique resolved names (excluding non-vascular and non-matching taxa):</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(df.uni$name.short.correct)</code></pre></div>
<p>There are 83,679 unique (non-vascular plant) taxon names:</p>
<p><strong>Database affiliations (<code>sPlot 2.1</code> and <code>TRY 3.0</code>).</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$sPlot2<span class="fl">.1</span>.TRY)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular</span>
<span class="st">plant taxon names per unique to, and shared between TRY (S) and sPlot (T).&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">S</th>
<th align="right">ST</th>
<th align="right">T</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">34105</td>
<td align="right">20414</td>
<td align="right">29158</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(df.uni$Manual.matching)</code></pre></div>
<pre><code>## 
## sPlot2b       x 
##    1637     490</code></pre>
<p>1,432 of those names were matched manually (see above). <strong>Taxonomic ranks:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$rank.short.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular</span>
<span class="st">plant taxon names per taxonomic rank.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">family</th>
<th align="right">genus</th>
<th align="right">higher</th>
<th align="right">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">152</td>
<td align="right">4343</td>
<td align="right">13</td>
<td align="right">79169</td>
</tr>
</tbody>
</table>
<p><strong>Taxonomic status:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni$Status.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular plant</span>
<span class="st">taxon names that correspond to `Accepted`, `Synonyms` or Unresolved species, respecively.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">Accepted</th>
<th align="right">Synonym</th>
<th align="right">Unresolved</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">69173</td>
<td align="right">9998</td>
<td align="right">4506</td>
</tr>
</tbody>
</table>
<h2 id="stats-for-the-corrected-names-in-splot-only">Stats for the corrected names in <code>sPlot</code> only:</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.uni.splot &lt;-<span class="st"> </span>df.uni %&gt;%
<span class="st">    </span>dplyr::<span class="kw">filter</span>(is.vascular.species ==<span class="st"> </span><span class="ot">TRUE</span>, !<span class="kw">is.na</span>(name.short.correct), df.uni$sPlot2<span class="fl">.1</span>.TRY!=<span class="st"> &quot;T&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>((df.uni.splot$name.short.correct))</code></pre></div>
<pre><code>## [1] 54519</code></pre>
<p><strong>Database affiliations</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni.splot$sPlot2<span class="fl">.1</span>.TRY)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular</span>
<span class="st">plant taxon names per unique to sPlot (S), and shared between TRY and sPlot (ST).&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">S</th>
<th align="right">ST</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">34105</td>
<td align="right">20414</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(df.uni.splot$Manual.matching)</code></pre></div>
<pre><code>## 
## sPlot2b       x 
##    1637     266</code></pre>
<p>266 uniquenames in sPlot were matched manually (see above).</p>
<p><strong>Taxonomic ranks:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni.splot$rank.short.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized)</span>
<span class="st">vascular plant taxon names per taxonomic rank.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">family</th>
<th align="right">genus</th>
<th align="right">higher</th>
<th align="right">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">133</td>
<td align="right">2447</td>
<td align="right">9</td>
<td align="right">51930</td>
</tr>
</tbody>
</table>
<p><strong>Taxonomic status:</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">t</span>(<span class="kw">table</span>(df.uni.splot$Status.correct)), <span class="dt">caption =</span> <span class="st">&quot;Number of (standardized) vascular</span>
<span class="st">plant taxon names that correspond to `Accepted`, `Synonyms` or Unresolved species, respecively.&quot;</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">Accepted</th>
<th align="right">Synonym</th>
<th align="right">Unresolved</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">46009</td>
<td align="right">5967</td>
<td align="right">2543</td>
</tr>
</tbody>
</table>
<p><strong>Number of families in sPlot</strong>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(df.uni.splot$family.correct))</code></pre></div>
<pre><code>## [1] 439</code></pre>
<p><strong>Done!</strong></p>
<hr />
<h1 id="r-settings"><code>R</code>-settings</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS: /usr/lib/openblas-base/libopenblas.so.0
## LAPACK: /usr/lib/lapack/liblapack.so.3.0
## 
## locale:
##  [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8    
##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8   
##  [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] bindrcpp_0.2      rmarkdown_1.6     rgbif_0.9.8      
##  [4] Taxonstand_2.0    pbapply_1.3-3     plyr_1.8.4       
##  [7] dplyr_0.7.2.9000  doParallel_1.0.10 iterators_1.0.8  
## [10] foreach_1.4.3     vegdata_0.9       foreign_0.8-69   
## [13] knitr_1.16        stringr_1.2.0    
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.12      highr_0.6         compiler_3.4.1   
##  [4] bindr_0.1         tools_3.4.1       digest_0.6.12    
##  [7] evaluate_0.10.1   jsonlite_1.5      tibble_1.3.3     
## [10] gtable_0.2.0      lattice_0.20-35   pkgconfig_2.0.1  
## [13] rlang_0.1.1.9000  yaml_2.1.14       xml2_1.1.1       
## [16] httr_1.2.1.9000   rgeos_0.3-23      rprojroot_1.2    
## [19] grid_3.4.1        glue_1.1.1        data.table_1.10.4
## [22] geoaxe_0.1.0      R6_2.2.2          oai_0.2.2        
## [25] XML_3.98-1.9      sp_1.2-5          whisker_0.3-2    
## [28] ggplot2_2.2.1     magrittr_1.5      backports_1.1.0  
## [31] htmltools_0.3.6   scales_0.4.1      codetools_0.2-15 
## [34] assertthat_0.2.0  colorspace_1.3-2  stringi_1.1.5    
## [37] lazyeval_0.2.0    munsell_0.4.3</code></pre>
<h1 id="references">References</h1>
<p>Boyle, B., Hopkins, N., Lu, Z., Garay, J.A.R., Mozzherin, D., Rees, T., Matasci, N., Narro, M.L., Piel, W.H., Mckay, S.J. &amp; others. (2013) The taxonomic name resolution service: An online tool for automated standardization of plant names. <em>BMC Bioinformatics</em>, <strong>14</strong>, 16.</p>
<p>Chase, M. &amp; Reveal, J. (2009) A phylogenetic classification of the land plants to accompany apgiii. <em>Botanical Journal of the Linnean Society</em>, <strong>161</strong>, 122–127.</p>
<p>Federhen, S. (2010) The NCBI Handbook [Internet]. (eds J. McEntyre &amp; J. Ostell) Available from: <a href="http://www.ncbi.nlm.nih.gov/guide/taxonomy/" class="uri">http://www.ncbi.nlm.nih.gov/guide/taxonomy/</a>; National Center for Biotechnology Information, Bethesda, MD, USA, [Accessed: 25 Oct 2011].</p>
<p>Flann, C. (2009) <em>Global Compositae Checklist. [Accessed 2 Apr 2013]</em>. Available from: <a href="http://compositae.landcareresearch.co.nz/" class="uri">http://compositae.landcareresearch.co.nz/</a>.</p>
<p>International legume database and information service. (2006) <em>[Internet, Accessed 21 Aug 2015]</em>. Available from: <a href="http://www.ildis.org/LegumeWeb" class="uri">http://www.ildis.org/LegumeWeb</a>.</p>
<p>iPlant Collaborative. (2015) <em>The Taxonomic Name Resolution Service [Internet]. Version 4.0 [Accessed: 20 Sep 2015]</em>. Available from: <a href="http://tnrs.iplantcollaborative.org/" class="uri">http://tnrs.iplantcollaborative.org/</a>; National Center for Biotechnology Information.</p>
<p>Missouri Botanical Garden. (2013) <em>Tropicos.org [Internet, Accessed 19 Dec 2014]</em>. Available from: <a href="http://www.tropicos.org" class="uri">http://www.tropicos.org</a>.</p>
<p>The Plant List. (2013) <em>[Internet]. Version 1.1. [Accessed: 19 Aug 2015].</em> Available from: <a href="http://www.theplantlist.org/" class="uri">http://www.theplantlist.org/</a>.</p>
<p>USDA, NRCS. (2012) <em>The Plants Database [Internet]. [Accessed: 17 Jan 2015]</em>. Available from: <a href="http://plants.usda.gov" class="uri">http://plants.usda.gov</a>; National Plant Data Team, Greensboro, NC, USA.</p>
<p>[1] <code>sPlot_14_4_2015_species</code> no longer exists, but will be used for documentation purposes here. This is the version of splot to generate the backbone for to match splot 2.0 to try 3.0. The following workflow describes the steps needed for generating a backbone for <code>sPlot_14_4_2015</code> and <code>TRY 3.0</code>. The backbone used in the third sPlot workshop contains the 7,701 additional (and/or different) species names in sPlot 2.1 (<code>sPlot_2015_07_19_species</code>) that had to be standardized in a ad-hoc fashion prior to the workshop.</p>
<p>[2] This species list only contains 122,901 species. The final backbone contains 7,701 additional (and/or different) species from from the ad-hoc version <code>sPlot_2015_07_19</code>.</p>
<p>[3] These were temporary files that were deleted after the name matching procedure.</p>
<p>[4] The full original <code>tnrs.tpl</code> table, including all 123,000+ species, has been overwritten, and now instead only contains the 121,000+ species from sPlot 2.1 and TRY 3.0.</p>
<p>[5] Note, the single temporary result files from TNRS are no longer available.</p>
<p>[6] Temporary file that no longer exists.</p>

</body>
</html>
